<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<meta name="date" content="2018-07-08" />

<title>AnnotationHub How-To’s</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">

</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="/Library/Frameworks/R.framework/Versions/3.5/Resources/library/BiocStyle/resources/html/bioconductor.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}

img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">seandavi(s12): Courses and Tutorials</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="setup.html">
    <span class="fa fa-cogs"></span>
     
    setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="ion ion-easel"></span>
     
    Slides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="motivation_for_R_slides.html">Motivation for using R</a>
    </li>
    <li>
      <a href="http://bit.ly/bioc_cshl_2018">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="MachineLearningSlides.html">Machine Learning hands-on</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
    Misc.
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="further_resources.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR">Source code for this site</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR/archive/master.zip">Download materials</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">AnnotationHub How-To’s</h1>
<h4 class="date"><em>8 July 2018</em></h4>

</div>

<h1>Contents</h1>
<div id="TOC">
<ul>
<li><a href="#accessing-genome-scale-data"><span class="toc-section-number">1</span> Accessing Genome-Scale Data</a><ul>
<li><a href="#non-model-organism-gene-annotations"><span class="toc-section-number">1.1</span> Non-model organism gene annotations</a></li>
<li><a href="#roadmap-epigenomics-project"><span class="toc-section-number">1.2</span> Roadmap Epigenomics Project</a></li>
<li><a href="#ensembl-gtf-and-fasta-files-for-txdb-gene-models-and-sequence-queries"><span class="toc-section-number">1.3</span> Ensembl GTF and FASTA files for TxDb gene models and sequence queries</a></li>
<li><a href="#liftover-to-map-between-genome-builds"><span class="toc-section-number">1.4</span> liftOver to map between genome builds</a></li>
<li><a href="#working-with-dbsnp-variants"><span class="toc-section-number">1.5</span> Working with dbSNP Variants</a></li>
</ul></li>
<li><a href="#sessioninfo"><span class="toc-section-number">2</span> sessionInfo</a></li>
</ul>
</div>

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<p><strong>Package</strong>: <em><a href="http://bioconductor.org/packages/AnnotationHub">AnnotationHub</a></em><br />
<strong>Authors</strong>: Martin Morgan [cre],
Marc Carlson [ctb],
Dan Tenenbaum [ctb],
Sonali Arora [ctb]<br />
<strong>Modified</strong>: Sun Jun 28 10:41:23 2015<br />
<strong>Compiled</strong>: Sun Jul 8 08:27:40 2018</p>
<div id="accessing-genome-scale-data" class="section level1">
<h1><span class="header-section-number">1</span> Accessing Genome-Scale Data</h1>
<div id="non-model-organism-gene-annotations" class="section level2">
<h2><span class="header-section-number">1.1</span> Non-model organism gene annotations</h2>
<p><em>Bioconductor</em> offers pre-built <code>org.*</code> annotation packages for model
organisms, with their use described in the
<a href="http://bioconductor.org/help/workflows/annotation/Annotation_Resources/#OrgDb">OrgDb</a>
section of the Annotation work flow. Here we discover available <code>OrgDb</code>
objects for less-model organisms</p>
<pre class="r"><code>library(AnnotationHub)
ah &lt;- AnnotationHub()</code></pre>
<pre><code>## snapshotDate(): 2018-04-30</code></pre>
<pre class="r"><code>query(ah, &quot;OrgDb&quot;)</code></pre>
<pre><code>## AnnotationHub with 1691 records
## # snapshotDate(): 2018-04-30 
## # $dataprovider: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/
## # $species: Escherichia coli, &#39;Caballeronia concitans&#39;, &#39;Chlorella vulgaris&#39;_C-169, &#39;Frankia cas...
## # $rdataclass: OrgDb
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH61768&quot;]]&#39; 
## 
##             title                                      
##   AH61768 | org.Ag.eg.db.sqlite                        
##   AH61769 | org.At.tair.db.sqlite                      
##   AH61770 | org.Bt.eg.db.sqlite                        
##   AH61771 | org.Cf.eg.db.sqlite                        
##   AH61772 | org.Gg.eg.db.sqlite                        
##   ...       ...                                        
##   AH63468 | org.Salmonella_typhimurium_LT2.eg.sqlite   
##   AH63469 | org.Acinetobacter_baumannii.eg.sqlite      
##   AH63470 | org.Acinetobacter_genomosp._2.eg.sqlite    
##   AH63471 | org.Acinetobacter_genomospecies_2.eg.sqlite
##   AH63472 | org.Bacterium_anitratum.eg.sqlite</code></pre>
<pre class="r"><code>orgdb &lt;- query(ah, &quot;OrgDb&quot;)[[1]] </code></pre>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/68514&#39;</code></pre>
<p>The object returned by AnnotationHub is directly usable with the
<code>select()</code> interface, e.g., to discover the available keytypes for
querying the object, the columns that these keytypes can map to, and
finally selecting the SYMBOL and GENENAME corresponding to the first 6
ENTREZIDs</p>
<pre class="r"><code>keytypes(orgdb)</code></pre>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot; &quot;ENTREZID&quot;     &quot;ENZYME&quot;      
##  [7] &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;ONTOLOGY&quot;    
## [13] &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PMID&quot;         &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UNIGENE&quot;     
## [19] &quot;UNIPROT&quot;</code></pre>
<pre class="r"><code>columns(orgdb)</code></pre>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot; &quot;ENTREZID&quot;     &quot;ENZYME&quot;      
##  [7] &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;ONTOLOGY&quot;    
## [13] &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PMID&quot;         &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UNIGENE&quot;     
## [19] &quot;UNIPROT&quot;</code></pre>
<pre class="r"><code>egid &lt;- head(keys(orgdb, &quot;ENTREZID&quot;))
select(orgdb, egid, c(&quot;SYMBOL&quot;, &quot;GENENAME&quot;), &quot;ENTREZID&quot;)</code></pre>
<pre><code>## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>##   ENTREZID          SYMBOL      GENENAME
## 1  1267437 AgaP_AGAP012606 AGAP012606-PA
## 2  1267439 AgaP_AGAP012559 AGAP012559-PA
## 3  1267440 AgaP_AGAP012558 AGAP012558-PA
## 4  1267447 AgaP_AGAP012586 AGAP012586-PA
## 5  1267450 AgaP_AGAP012834 AGAP012834-PA
## 6  1267459 AgaP_AGAP012589 AGAP012589-PA</code></pre>
</div>
<div id="roadmap-epigenomics-project" class="section level2">
<h2><span class="header-section-number">1.2</span> Roadmap Epigenomics Project</h2>
<p>All Roadmap Epigenomics files are hosted
<a href="http://egg2.wustl.edu/roadmap/data/byFileType/">here</a>. If one had to
download these files on their own, one would navigate through the web
interface to find useful files, then use something like the following
<em>R</em> code.</p>
<pre class="r"><code>url &lt;- &quot;http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/broadPeak/E001-H3K4me1.broadPeak.gz&quot;
filename &lt;-  basename(url)
download.file(url, destfile=filename)
if (file.exists(filename))
   data &lt;- import(filename, format=&quot;bed&quot;)</code></pre>
<p>This would have to be repeated for all files, and the onus would lie
on the user to identify, download, import, and manage the local disk
location of these files.</p>
<p><em><a href="http://bioconductor.org/packages/AnnotationHub">AnnotationHub</a></em> reduces this task to just a few lines of <em>R</em> code</p>
<pre class="r"><code>library(AnnotationHub)
ah = AnnotationHub()</code></pre>
<pre><code>## snapshotDate(): 2018-04-30</code></pre>
<pre class="r"><code>epiFiles &lt;- query(ah, &quot;EpigenomeRoadMap&quot;)</code></pre>
<p>A look at the value returned by <code>epiFiles</code> shows us that
18248 roadmap resources are available via
<em><a href="http://bioconductor.org/packages/AnnotationHub">AnnotationHub</a></em>. Additional information about
the files is also available, e.g., where the files came from
(dataprovider), genome, species, sourceurl, sourcetypes.</p>
<pre class="r"><code>epiFiles</code></pre>
<pre><code>## AnnotationHub with 18248 records
## # snapshotDate(): 2018-04-30 
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: BigWigFile, GRanges, data.frame
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH28856&quot;]]&#39; 
## 
##             title                                 
##   AH28856 | E001-H3K4me1.broadPeak.gz             
##   AH28857 | E001-H3K4me3.broadPeak.gz             
##   AH28858 | E001-H3K9ac.broadPeak.gz              
##   AH28859 | E001-H3K9me3.broadPeak.gz             
##   AH28860 | E001-H3K27me3.broadPeak.gz            
##   ...       ...                                   
##   AH49540 | E058_mCRF_FractionalMethylation.bigwig
##   AH49541 | E059_mCRF_FractionalMethylation.bigwig
##   AH49542 | E061_mCRF_FractionalMethylation.bigwig
##   AH49543 | E081_mCRF_FractionalMethylation.bigwig
##   AH49544 | E082_mCRF_FractionalMethylation.bigwig</code></pre>
<p>A good sanity check to ensure that we have files only from the Roadmap Epigenomics
project is to check that all the files in the returned smaller hub object
come from <em>Homo sapiens</em> and the hg19 genome</p>
<pre class="r"><code>unique(epiFiles$species)</code></pre>
<pre><code>## [1] &quot;Homo sapiens&quot;</code></pre>
<pre class="r"><code>unique(epiFiles$genome)</code></pre>
<pre><code>## [1] &quot;hg19&quot;</code></pre>
<p>Broadly, one can get an idea of the different files from this project
looking at the sourcetype</p>
<pre class="r"><code>table(epiFiles$sourcetype)</code></pre>
<pre><code>## 
##    BED BigWig    GTF    tab    Zip 
##   8298   9932      3      1     14</code></pre>
<p>To get a more descriptive idea of these different files one can use:</p>
<pre class="r"><code>sort(table(epiFiles$description), decreasing=TRUE)</code></pre>
<pre><code>## 
##                       Bigwig File containing -log10(p-value) signal tracks from EpigenomeRoadMap Project 
##                                                                                                     6881 
##                       Bigwig File containing fold enrichment signal tracks from EpigenomeRoadMap Project 
##                                                                                                     2947 
##                          Narrow ChIP-seq peaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                                                     2894 
##                           Broad ChIP-seq peaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                                                     2534 
##                          Gapped ChIP-seq peaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                                                     2534 
##                              Narrow DNasePeaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                                                      131 
##                                           15 state chromatin segmentations from EpigenomeRoadMap Project 
##                                                                                                      127 
##      Broad domains on enrichment for DNase-seq for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                                                       78 
##                                         RRBS fractional methylation calls from EpigenomeRoadMap Project  
##                                                                                                       51 
##                       Whole genome bisulphite fractional methylation calls from EpigenomeRoadMap Project 
##                                                                                                       37 
##                               MeDIP/MRE(mCRF) fractional methylation calls from EpigenomeRoadMap Project 
##                                                                                                       16 
## GencodeV10 gene/transcript coordinates and annotations corresponding to hg19 version of the human genome 
##                                                                                                        3 
##                                       RNA-seq read count matrix for intronic protein-coding RNA elements 
##                                                                                                        2 
##                                                      RNA-seq read counts matrix for ribosomal gene exons 
##                                                                                                        2 
##                                                          RPKM expression matrix for ribosomal gene exons 
##                                                                                                        2 
##                                                                    Metadata for EpigenomeRoadMap Project 
##                                                                                                        1 
##                                                           RNA-seq read counts matrix for non-coding RNAs 
##                                                                                                        1 
##                                                      RNA-seq read counts matrix for protein coding exons 
##                                                                                                        1 
##                                                      RNA-seq read counts matrix for protein coding genes 
##                                                                                                        1 
##                                                           RNA-seq read counts matrix for ribosomal genes 
##                                                                                                        1 
##                                                               RPKM expression matrix for non-coding RNAs 
##                                                                                                        1 
##                                                          RPKM expression matrix for protein coding exons 
##                                                                                                        1 
##                                                          RPKM expression matrix for protein coding genes 
##                                                                                                        1 
##                                                                RPKM expression matrix for ribosomal RNAs 
##                                                                                                        1</code></pre>
<p>The ‘metadata’ provided by the Roadmap Epigenomics Project is also
available. Note that the information displayed about a hub with a
single resource is quite different from the information displayed when
the hub references more than one resource.</p>
<pre class="r"><code>metadata.tab &lt;- query(ah , c(&quot;EpigenomeRoadMap&quot;, &quot;Metadata&quot;))
metadata.tab</code></pre>
<pre><code>## AnnotationHub with 1 record
## # snapshotDate(): 2018-04-30 
## # names(): AH41830
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: data.frame
## # $rdatadateadded: 2015-05-11
## # $title: EID_metadata.tab
## # $description: Metadata for EpigenomeRoadMap Project
## # $taxonomyid: 9606
## # $genome: hg19
## # $sourcetype: tab
## # $sourceurl: http://egg2.wustl.edu/roadmap/data/byFileType/metadata/EID_metadata.tab
## # $sourcesize: 18035
## # $tags: c(&quot;EpigenomeRoadMap&quot;, &quot;Metadata&quot;) 
## # retrieve record with &#39;object[[&quot;AH41830&quot;]]&#39;</code></pre>
<p>So far we have been exploring information about resources, without
downloading the resource to a local cache and importing it into R.
One can retrieve the resource using <code>[[</code> as indicated at the
end of the show method</p>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/47270&#39;</code></pre>
<pre class="r"><code>metadata.tab &lt;- ah[[&quot;AH41830&quot;]]</code></pre>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/47270&#39;</code></pre>
<p>The metadata.tab file is returned as a <em>data.frame</em>. The first 6 rows
of the first 5 columns are shown here:</p>
<pre class="r"><code>metadata.tab[1:6, 1:5]</code></pre>
<pre><code>##    EID    GROUP   COLOR          MNEMONIC                                   STD_NAME
## 1 E001      ESC #924965            ESC.I3                                ES-I3 Cells
## 2 E002      ESC #924965           ESC.WA7                               ES-WA7 Cells
## 3 E003      ESC #924965            ESC.H1                                   H1 Cells
## 4 E004 ES-deriv #4178AE ESDR.H1.BMP4.MESO H1 BMP4 Derived Mesendoderm Cultured Cells
## 5 E005 ES-deriv #4178AE ESDR.H1.BMP4.TROP H1 BMP4 Derived Trophoblast Cultured Cells
## 6 E006 ES-deriv #4178AE       ESDR.H1.MSC          H1 Derived Mesenchymal Stem Cells</code></pre>
<p>One can keep constructing different queries using multiple arguments to
trim down these 18248 to get the files one wants.
For example, to get the ChIP-Seq files for consolidated epigenomes,
one could use</p>
<pre class="r"><code>bpChipEpi &lt;- query(ah , c(&quot;EpigenomeRoadMap&quot;, &quot;broadPeak&quot;, &quot;chip&quot;, &quot;consolidated&quot;))</code></pre>
<p>To get all the bigWig signal files, one can query the hub using</p>
<pre class="r"><code>allBigWigFiles &lt;- query(ah, c(&quot;EpigenomeRoadMap&quot;, &quot;BigWig&quot;))</code></pre>
<p>To access the 15 state chromatin segmentations, one can use</p>
<pre class="r"><code>seg &lt;- query(ah, c(&quot;EpigenomeRoadMap&quot;, &quot;segmentations&quot;))</code></pre>
<p>If one is interested in getting all the files related to one sample</p>
<pre class="r"><code>E126 &lt;- query(ah , c(&quot;EpigenomeRoadMap&quot;, &quot;E126&quot;, &quot;H3K4ME2&quot;))
E126</code></pre>
<pre><code>## AnnotationHub with 6 records
## # snapshotDate(): 2018-04-30 
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: BigWigFile, GRanges
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH29817&quot;]]&#39; 
## 
##             title                                  
##   AH29817 | E126-H3K4me2.broadPeak.gz              
##   AH30868 | E126-H3K4me2.narrowPeak.gz             
##   AH31801 | E126-H3K4me2.gappedPeak.gz             
##   AH32990 | E126-H3K4me2.fc.signal.bigwig          
##   AH34022 | E126-H3K4me2.pval.signal.bigwig        
##   AH40177 | E126-H3K4me2.imputed.pval.signal.bigwig</code></pre>
<p>Hub resources can also be selected using <code>$</code>, <code>subset()</code>, and
<code>display()</code>; see the main
<a href="AnnotationHub.html"><em>AnnotationHub</em> vignette</a> for additional detail.</p>
<p>Hub resources are imported as the appropriate <em>Bioconductor</em> object
for use in further analysis. For example, peak files are returned as
<em>GRanges</em> objects.</p>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/35257&#39;</code></pre>
<pre class="r"><code>peaks &lt;- E126[[&#39;AH29817&#39;]]</code></pre>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/35257&#39;</code></pre>
<pre class="r"><code>seqinfo(peaks)</code></pre>
<pre><code>## Seqinfo object with 93 sequences (1 circular) from hg19 genome:
##   seqnames       seqlengths isCircular genome
##   chr1            249250621      FALSE   hg19
##   chr2            243199373      FALSE   hg19
##   chr3            198022430      FALSE   hg19
##   chr4            191154276      FALSE   hg19
##   chr5            180915260      FALSE   hg19
##   ...                   ...        ...    ...
##   chrUn_gl000245      36651      FALSE   hg19
##   chrUn_gl000246      38154      FALSE   hg19
##   chrUn_gl000247      36422      FALSE   hg19
##   chrUn_gl000248      39786      FALSE   hg19
##   chrUn_gl000249      38502      FALSE   hg19</code></pre>
<p>BigWig files are returned as <em>BigWigFile</em> objects. A <em>BigWigFile</em> is a
reference to a file on disk; the data in the file can be read in using
<code>rtracklayer::import()</code>, perhaps querying these large files for
particular genomic regions of interest as described on the help page
<code>?import.bw</code>.</p>
<p>Each record inside <em><a href="http://bioconductor.org/packages/AnnotationHub">AnnotationHub</a></em> is associated with a
unique identifier. Most <em>GRanges</em> objects returned by
<em><a href="http://bioconductor.org/packages/AnnotationHub">AnnotationHub</a></em> contain the unique AnnotationHub identifier of
the resource from which the <em>GRanges</em> is derived. This can come handy
when working with the <em>GRanges</em> object for a while, and additional
information about the object (e.g., the name of the file in the cache,
or the original sourceurl for the data underlying the resource) that
is being worked with.</p>
<pre class="r"><code>metadata(peaks)</code></pre>
<pre><code>## $AnnotationHubName
## [1] &quot;AH29817&quot;
## 
## $`File Name`
## [1] &quot;E126-H3K4me2.broadPeak.gz&quot;
## 
## $`Data Source`
## [1] &quot;http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/broadPeak/E126-H3K4me2.broadPeak.gz&quot;
## 
## $Provider
## [1] &quot;BroadInstitute&quot;
## 
## $Organism
## [1] &quot;Homo sapiens&quot;
## 
## $`Taxonomy ID`
## [1] 9606</code></pre>
<pre class="r"><code>ah[metadata(peaks)$AnnotationHubName]$sourceurl</code></pre>
<pre><code>## [1] &quot;http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/broadPeak/E126-H3K4me2.broadPeak.gz&quot;</code></pre>
</div>
<div id="ensembl-gtf-and-fasta-files-for-txdb-gene-models-and-sequence-queries" class="section level2">
<h2><span class="header-section-number">1.3</span> Ensembl GTF and FASTA files for TxDb gene models and sequence queries</h2>
<p><em>Bioconductor</em> represents gene models using ‘transcript’
databases. These are available via packages such as
<em><a href="http://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg38.knownGene">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>
or can be constructed using functions such as
<em><a href="http://bioconductor.org/packages/GenomicFeatures">GenomicFeatures</a></em>::<code>makeTxDbFromBiomart()</code>.</p>
<p><em>AnnotationHub</em> provides an easy way to work with gene models
published by Ensembl. Let’s see what Ensembl’s Release-80 has in terms
of data for pufferfish, <em>Takifugu rubripes</em>.</p>
<pre class="r"><code>query(ah, c(&quot;Takifugu&quot;, &quot;release-80&quot;))</code></pre>
<pre><code>## AnnotationHub with 7 records
## # snapshotDate(): 2018-04-30 
## # $dataprovider: Ensembl
## # $species: Takifugu rubripes
## # $rdataclass: FaFile, GRanges
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH47101&quot;]]&#39; 
## 
##             title                                     
##   AH47101 | Takifugu_rubripes.FUGU4.80.gtf            
##   AH47475 | Takifugu_rubripes.FUGU4.cdna.all.fa       
##   AH47476 | Takifugu_rubripes.FUGU4.dna_rm.toplevel.fa
##   AH47477 | Takifugu_rubripes.FUGU4.dna_sm.toplevel.fa
##   AH47478 | Takifugu_rubripes.FUGU4.dna.toplevel.fa   
##   AH47479 | Takifugu_rubripes.FUGU4.ncrna.fa          
##   AH47480 | Takifugu_rubripes.FUGU4.pep.all.fa</code></pre>
<p>We see that there is a GTF file descrbing gene models, as well as
various DNA sequences. Let’s retrieve the GTF and top-level DNA
sequence files. The GTF file is imported as a <em>GRanges</em> instance, the
DNA sequence as a compressed, indexed Fasta file</p>
<pre class="r"><code>gtf &lt;- ah[[&quot;AH47101&quot;]]</code></pre>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/52579&#39;</code></pre>
<pre class="r"><code>dna &lt;- ah[[&quot;AH47477&quot;]]</code></pre>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/53323&#39;
##     &#39;/Users/sdavis2//.AnnotationHub/53324&#39;</code></pre>
<pre class="r"><code>head(gtf, 3)</code></pre>
<pre><code>## GRanges object with 3 ranges and 19 metadata columns:
##         seqnames      ranges strand |   source       type     score     phase            gene_id
##            &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;        &lt;character&gt;
##   [1] scaffold_1 10422-11354      - |  ensembl       gene      &lt;NA&gt;      &lt;NA&gt; ENSTRUG00000003702
##   [2] scaffold_1 10422-11354      - |  ensembl transcript      &lt;NA&gt;      &lt;NA&gt; ENSTRUG00000003702
##   [3] scaffold_1 10422-11354      - |  ensembl       exon      &lt;NA&gt;      &lt;NA&gt; ENSTRUG00000003702
##       gene_version gene_source   gene_biotype      transcript_id transcript_version
##          &lt;numeric&gt; &lt;character&gt;    &lt;character&gt;        &lt;character&gt;          &lt;numeric&gt;
##   [1]            1     ensembl protein_coding               &lt;NA&gt;               &lt;NA&gt;
##   [2]            1     ensembl protein_coding ENSTRUT00000008740                  1
##   [3]            1     ensembl protein_coding ENSTRUT00000008740                  1
##       transcript_source transcript_biotype exon_number            exon_id exon_version  protein_id
##             &lt;character&gt;        &lt;character&gt;   &lt;numeric&gt;        &lt;character&gt;    &lt;numeric&gt; &lt;character&gt;
##   [1]              &lt;NA&gt;               &lt;NA&gt;        &lt;NA&gt;               &lt;NA&gt;         &lt;NA&gt;        &lt;NA&gt;
##   [2]           ensembl     protein_coding        &lt;NA&gt;               &lt;NA&gt;         &lt;NA&gt;        &lt;NA&gt;
##   [3]           ensembl     protein_coding           1 ENSTRUE00000055472            1        &lt;NA&gt;
##       protein_version   gene_name transcript_name
##             &lt;numeric&gt; &lt;character&gt;     &lt;character&gt;
##   [1]            &lt;NA&gt;        &lt;NA&gt;            &lt;NA&gt;
##   [2]            &lt;NA&gt;        &lt;NA&gt;            &lt;NA&gt;
##   [3]            &lt;NA&gt;        &lt;NA&gt;            &lt;NA&gt;
##   -------
##   seqinfo: 2056 sequences (1 circular) from FUGU4 genome; no seqlengths</code></pre>
<pre class="r"><code>dna</code></pre>
<pre><code>## class: FaFile 
## path: /Users/sdavis2//.AnnotationHub/53323
## index: /Users/sdavis2//.AnnotationHub/53324
## isOpen: FALSE 
## yieldSize: NA</code></pre>
<pre class="r"><code>head(seqlevels(dna))</code></pre>
<pre><code>## [1] &quot;scaffold_1&quot; &quot;scaffold_2&quot; &quot;scaffold_3&quot; &quot;scaffold_4&quot; &quot;scaffold_5&quot; &quot;scaffold_6&quot;</code></pre>
<p>Let’s identify the 25 longest DNA sequences, and keep just the
annotations on these scaffolds.</p>
<pre class="r"><code>keep &lt;- names(tail(sort(seqlengths(dna)), 25))
gtf_subset &lt;- gtf[seqnames(gtf) %in% keep]</code></pre>
<p>It is trivial to make a TxDb instance of this subset (or of the entire
gtf)</p>
<pre class="r"><code>library(GenomicFeatures)         # for makeTxDbFromGRanges
txdb &lt;- makeTxDbFromGRanges(gtf_subset)</code></pre>
<pre><code>## Warning in .get_cds_IDX(type, phase): The &quot;phase&quot; metadata column contains non-NA values for features of type stop_codon. This
##   information was ignored.</code></pre>
<p>and to use that in conjunction with the DNA sequences, e.g., to find
exon sequences of all annotated genes.</p>
<pre class="r"><code>library(Rsamtools)               # for getSeq,FaFile-method
exons &lt;- exons(txdb)
length(exons)</code></pre>
<pre><code>## [1] 66219</code></pre>
<pre class="r"><code>getSeq(dna, exons)</code></pre>
<pre><code>##   A DNAStringSet instance of length 66219
##         width seq                                                               names               
##     [1]    72 ATGGCCTATCAGTTGTACAGGAATACCACTC...CCTGCAGGAGAGTCTGGACGAGCTTATCCAG scaffold_1
##     [2]   105 ACTCAGCAGATCACCCCTCAGCTGGCTCTCC...TAATCGTGTCCGCAACCGTGTGAACTTCAGG scaffold_1
##     [3]   156 GGTTCTCTCAACACCTACCGGTTCTGTGACA...GAGCGAATCCATGCAAAACAAACTGGATAAA scaffold_1
##     [4]    88 CAAACCAATCTCCTCGCTGTCTCTTCTCGTT...CATCAGCCAGAGGGACGGATCATCTCAGGTT scaffold_1
##     [5]   271 AGACGAGATGAGTGAGGACGCATTCAACGCC...AACACAGTGTGGAGACTTCAGAGGACGCCAC scaffold_1
##     ...   ... ...
## [66215]    67 ACGACTGGATGACAACATCAGGACCGTGGTA...TCAGACCAATGTGGGTCAGGATGGCAGACAG scaffold_9
## [66216]    50 TCTTTGGCTAATATTGACGATGTGGTAAACAAGATTCGTCTGAAGATTCG                scaffold_9
## [66217]    81 GTATTTCCCAGCCAAGACCCGCTGGACAGGG...ATACATCAACACACTGTTTCCCACCGAGCAG scaffold_9
## [66218]    87 ATGATGGAGGATGAAGAATTTGAATTTGCGG...GACCCCAGAGGTGCAGCTAGCAATTGAACAG scaffold_9
## [66219]   213 GACGACATCCTCGTGTGGGGCCGCTCTAGGG...GACAGCTGCTGTTCGCCTGTTTCCCCCCCCC scaffold_9</code></pre>
<p>There is a one-to-one mapping between the genomic ranges contained in
<code>exons</code> and the DNA sequences returned by <code>getSeq()</code>.</p>
<p>Some difficulties arise when working with this partly assembled genome
that require more advanced GenomicRanges skills, see the
<em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> vignettes, especially “<em>GenomicRanges</em>
HOWTOs” and “An Introduction to <em>GenomicRanges</em>”.</p>
</div>
<div id="liftover-to-map-between-genome-builds" class="section level2">
<h2><span class="header-section-number">1.4</span> liftOver to map between genome builds</h2>
<p>Suppose we wanted to lift features from one genome build to another,
e.g., because annotations were generated for hg19 but our experimental
analysis used hg18. We know that UCSC provides ‘liftover’ files for
mapping between genome builds.</p>
<p>In this example, we will take our broad Peak <em>GRanges</em> from E126 which
comes from the ‘hg19’ genome, and lift over these features to their
‘hg38’ coordinates.</p>
<pre class="r"><code>chainfiles &lt;- query(ah , c(&quot;hg38&quot;, &quot;hg19&quot;, &quot;chainfile&quot;))
chainfiles</code></pre>
<pre><code>## AnnotationHub with 2 records
## # snapshotDate(): 2018-04-30 
## # $dataprovider: UCSC
## # $species: Homo sapiens
## # $rdataclass: ChainFile
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH14108&quot;]]&#39; 
## 
##             title                   
##   AH14108 | hg38ToHg19.over.chain.gz
##   AH14150 | hg19ToHg38.over.chain.gz</code></pre>
<p>We are interested in the file that lifts over features from hg19 to
hg38 so lets download that using</p>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/18245&#39;</code></pre>
<pre class="r"><code>chain &lt;- chainfiles[[&#39;AH14150&#39;]]</code></pre>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     &#39;/Users/sdavis2//.AnnotationHub/18245&#39;</code></pre>
<pre class="r"><code>chain</code></pre>
<pre><code>## Chain of length 25
## names(25): chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 ... chr18 chr19 chr20 chr21 chr22 chrX chrY chrM</code></pre>
<p>Perform the liftOver operation using <code>rtracklayer::liftOver()</code>:</p>
<pre class="r"><code>library(rtracklayer)
gr38 &lt;- liftOver(peaks, chain)</code></pre>
<p>This returns a <em>GRangeslist</em>; update the genome of the result to get
the final result</p>
<pre class="r"><code>genome(gr38) &lt;- &quot;hg38&quot;
gr38</code></pre>
<pre><code>## GRangesList object of length 153266:
## [[1]] 
## GRanges object with 1 range and 5 metadata columns:
##       seqnames            ranges strand |        name     score signalValue    pValue    qValue
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##   [1]     chr1 28667912-28670147      * |      Rank_1       189    10.55845  22.01316  18.99911
## 
## [[2]] 
## GRanges object with 1 range and 5 metadata columns:
##       seqnames            ranges strand |   name score signalValue   pValue   qValue
##   [1]     chr4 54090990-54092984      * | Rank_2   188     8.11483 21.80441 18.80662
## 
## [[3]] 
## GRanges object with 1 range and 5 metadata columns:
##       seqnames            ranges strand |   name score signalValue   pValue   qValue
##   [1]    chr14 75293392-75296621      * | Rank_3   180     8.89834 20.97714 18.02816
## 
## ...
## &lt;153263 more elements&gt;
## -------
## seqinfo: 23 sequences from hg38 genome; no seqlengths</code></pre>
</div>
<div id="working-with-dbsnp-variants" class="section level2">
<h2><span class="header-section-number">1.5</span> Working with dbSNP Variants</h2>
<p>One may also be interested in working with common germline variants with
evidence of medical interest. This information is available at
<a href="https://www.ncbi.nlm.nih.gov/variation/docs/human_variation_vcf/">NCBI</a>.</p>
<p>Query the dbDNP files in the hub:</p>
<p>This returns a <em>VcfFile</em> which can be read in using <code>r Biocpkg(&quot;VariantAnnotation&quot;)</code>; because VCF files can be large, <code>readVcf()</code>
supports several strategies for importing only relevant parts of the file
(e.g., particular genomic locations, particular features of the variants), see
<code>?readVcf</code> for additional information.</p>
<pre class="r"><code>variants &lt;- readVcf(vcf, genome=&quot;hg19&quot;)
variants</code></pre>
<pre><code>## class: CollapsedVCF 
## dim: 111138 0 
## rowRanges(vcf):
##   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER
## info(vcf):
##   DataFrame with 58 columns: RS, RSPOS, RV, VP, GENEINFO, dbSNPBuildID, SAO, SSR, WGT, VC, PM, T...
## info(header(vcf)):
##                 Number Type    Description                                                         
##    RS           1      Integer dbSNP ID (i.e. rs number)                                           
##    RSPOS        1      Integer Chr position reported in dbSNP                                      
##    RV           0      Flag    RS orientation is reversed                                          
##    VP           1      String  Variation Property.  Documentation is at ftp://ftp.ncbi.nlm.nih.g...
##    GENEINFO     1      String  Pairs each of gene symbol:gene id.  The gene symbol and id are de...
##    dbSNPBuildID 1      Integer First dbSNP Build for RS                                            
##    SAO          1      Integer Variant Allele Origin: 0 - unspecified, 1 - Germline, 2 - Somatic...
##    SSR          1      Integer Variant Suspect Reason Codes (may be more than one value added to...
##    WGT          1      Integer Weight, 00 - unmapped, 1 - weight 1, 2 - weight 2, 3 - weight 3 o...
##    VC           1      String  Variation Class                                                     
##    PM           0      Flag    Variant is Precious(Clinical,Pubmed Cited)                          
##    TPA          0      Flag    Provisional Third Party Annotation(TPA) (currently rs from PHARMG...
##    PMC          0      Flag    Links exist to PubMed Central article                               
##    S3D          0      Flag    Has 3D structure - SNP3D table                                      
##    SLO          0      Flag    Has SubmitterLinkOut - From SNP-&gt;SubSNP-&gt;Batch.link_out             
##    NSF          0      Flag    Has non-synonymous frameshift A coding region variation where one...
##    NSM          0      Flag    Has non-synonymous missense A coding region variation where one a...
##    NSN          0      Flag    Has non-synonymous nonsense A coding region variation where one a...
##    REF          0      Flag    Has reference A coding region variation where one allele in the s...
##    SYN          0      Flag    Has synonymous A coding region variation where one allele in the ...
##    U3           0      Flag    In 3&#39; UTR Location is in an untranslated region (UTR). FxnCode = 53 
##    U5           0      Flag    In 5&#39; UTR Location is in an untranslated region (UTR). FxnCode = 55 
##    ASS          0      Flag    In acceptor splice site FxnCode = 73                                
##    DSS          0      Flag    In donor splice-site FxnCode = 75                                   
##    INT          0      Flag    In Intron FxnCode = 6                                               
##    R3           0      Flag    In 3&#39; gene region FxnCode = 13                                      
##    R5           0      Flag    In 5&#39; gene region FxnCode = 15                                      
##    OTH          0      Flag    Has other variant with exactly the same set of mapped positions o...
##    CFL          0      Flag    Has Assembly conflict. This is for weight 1 and 2 variant that ma...
##    ASP          0      Flag    Is Assembly specific. This is set if the variant only maps to one...
##    MUT          0      Flag    Is mutation (journal citation, explicit fact): a low frequency va...
##    VLD          0      Flag    Is Validated.  This bit is set if the variant has 2+ minor allele...
##    G5A          0      Flag    &gt;5% minor allele frequency in each and all populations              
##    G5           0      Flag    &gt;5% minor allele frequency in 1+ populations                        
##    HD           0      Flag    Marker is on high density genotyping kit (50K density or greater)...
##    GNO          0      Flag    Genotypes available. The variant has individual genotype (in SubI...
##    KGPhase1     0      Flag    1000 Genome phase 1 (incl. June Interim phase 1)                    
##    KGPhase3     0      Flag    1000 Genome phase 3                                                 
##    CDA          0      Flag    Variation is interrogated in a clinical diagnostic assay            
##    LSD          0      Flag    Submitted from a locus-specific database                            
##    MTP          0      Flag    Microattribution/third-party annotation(TPA:GWAS,PAGE)              
##    OM           0      Flag    Has OMIM/OMIA                                                       
##    NOC          0      Flag    Contig allele not present in variant allele list. The reference s...
##    WTD          0      Flag    Is Withdrawn by submitter If one member ss is withdrawn by submit...
##    NOV          0      Flag    Rs cluster has non-overlapping allele sets. True when rs set has ...
##    CAF          .      String  An ordered, comma delimited list of allele frequencies based on 1...
##    COMMON       1      Integer RS is a common SNP.  A common SNP is one that has at least one 10...
##    CLNHGVS      .      String  Variant names from HGVS.    The order of these variants correspon...
##    CLNALLE      .      Integer Variant alleles from REF or ALT columns.  0 is REF, 1 is the firs...
##    CLNSRC       .      String  Variant Clinical Chanels                                            
##    CLNORIGIN    .      String  Allele Origin. One or more of the following values may be added: ...
##    CLNSRCID     .      String  Variant Clinical Channel IDs                                        
##    CLNSIG       .      String  Variant Clinical Significance, 0 - Uncertain significance, 1 - no...
##    CLNDSDB      .      String  Variant disease database name                                       
##    CLNDSDBID    .      String  Variant disease database ID                                         
##    CLNDBN       .      String  Variant disease name                                                
##    CLNREVSTAT   .      String  no_assertion - No assertion provided, no_criteria - No assertion ...
##    CLNACC       .      String  Variant Accession and Versions                                      
## geno(vcf):
##   SimpleList of length 0:</code></pre>
<p><code>rowRanges()</code> returns information from the CHROM, POS and ID fields of the VCF
file, represented as a <em>GRanges</em> instance</p>
<pre class="r"><code>rowRanges(variants)</code></pre>
<pre><code>## GRanges object with 111138 ranges and 5 metadata columns:
##               seqnames    ranges strand | paramRangeID            REF                ALT      QUAL
##                  &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt; &lt;DNAStringSetList&gt; &lt;numeric&gt;
##   rs786201005        1   1014143      * |         &lt;NA&gt;              C                  T      &lt;NA&gt;
##   rs672601345        1   1014316      * |         &lt;NA&gt;              C                 CG      &lt;NA&gt;
##   rs672601312        1   1014359      * |         &lt;NA&gt;              G                  T      &lt;NA&gt;
##   rs115173026        1   1020217      * |         &lt;NA&gt;              G                  T      &lt;NA&gt;
##   rs201073369        1   1020239      * |         &lt;NA&gt;              G                  C      &lt;NA&gt;
##           ...      ...       ...    ... .          ...            ...                ...       ...
##   rs527236200       MT     15943      * |         &lt;NA&gt;              T                  C      &lt;NA&gt;
##   rs118203890       MT     15950      * |         &lt;NA&gt;              G                  A      &lt;NA&gt;
##   rs199474700       MT     15965      * |         &lt;NA&gt;              A                  G      &lt;NA&gt;
##   rs199474701       MT     15967      * |         &lt;NA&gt;              G                  A      &lt;NA&gt;
##   rs199474699       MT     15990      * |         &lt;NA&gt;              C                  T      &lt;NA&gt;
##                    FILTER
##               &lt;character&gt;
##   rs786201005           .
##   rs672601345           .
##   rs672601312           .
##   rs115173026           .
##   rs201073369           .
##           ...         ...
##   rs527236200           .
##   rs118203890           .
##   rs199474700           .
##   rs199474701           .
##   rs199474699           .
##   -------
##   seqinfo: 25 sequences from hg19 genome; no seqlengths</code></pre>
<p>Note that the broadPeaks files follow the UCSC chromosome naming convention,
and the vcf data follows the NCBI style of chromosome naming convention.
To bring these ranges in the same chromosome
naming convention (ie UCSC), we would use</p>
<pre class="r"><code>seqlevelsStyle(variants) &lt;-seqlevelsStyle(peaks)</code></pre>
<p>And then finally to find which variants overlap these broadPeaks we would use:</p>
<pre class="r"><code>overlap &lt;- findOverlaps(variants, peaks)
overlap</code></pre>
<pre><code>## Hits object with 10904 hits and 0 metadata columns:
##           queryHits subjectHits
##           &lt;integer&gt;   &lt;integer&gt;
##       [1]        35       20333
##       [2]        36       20333
##       [3]        37       20333
##       [4]        38       20333
##       [5]        41        7733
##       ...       ...         ...
##   [10900]    110761       21565
##   [10901]    110762       21565
##   [10902]    110763       21565
##   [10903]    110764       21565
##   [10904]    110765       21565
##   -------
##   queryLength: 111138 / subjectLength: 153266</code></pre>
<p>Some insight into how these results can be interpretted comes from
looking a particular peak, e.g., the 3852nd peak</p>
<pre class="r"><code>idx &lt;- subjectHits(overlap) == 3852
overlap[idx]</code></pre>
<pre><code>## Hits object with 39 hits and 0 metadata columns:
##        queryHits subjectHits
##        &lt;integer&gt;   &lt;integer&gt;
##    [1]    102896        3852
##    [2]    102897        3852
##    [3]    102898        3852
##    [4]    102899        3852
##    [5]    102900        3852
##    ...       ...         ...
##   [35]    102930        3852
##   [36]    102931        3852
##   [37]    102932        3852
##   [38]    102933        3852
##   [39]    102934        3852
##   -------
##   queryLength: 111138 / subjectLength: 153266</code></pre>
<p>There are three variants overlapping this peak; the coordinates of the
peak and the overlapping variants are</p>
<pre class="r"><code>peaks[3852]</code></pre>
<pre><code>## GRanges object with 1 range and 5 metadata columns:
##       seqnames            ranges strand |        name     score signalValue    pValue    qValue
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##   [1]    chr22 50622494-50626143      * |   Rank_3852        79     6.06768  10.18943   7.99818
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<pre class="r"><code>rowRanges(variants)[queryHits(overlap[idx])]</code></pre>
<pre><code>## GRanges object with 39 ranges and 5 metadata columns:
##               seqnames    ranges strand | paramRangeID            REF                ALT      QUAL
##                  &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt; &lt;DNAStringSetList&gt; &lt;numeric&gt;
##     rs6151429    chr22  50625049      * |         &lt;NA&gt;              T                  C      &lt;NA&gt;
##     rs6151428    chr22  50625182      * |         &lt;NA&gt;              C                A,T      &lt;NA&gt;
##   rs774153480    chr22  50625182      * |         &lt;NA&gt;              C           CG,CGGGG      &lt;NA&gt;
##   rs199476388    chr22  50625204      * |         &lt;NA&gt;              A                C,G      &lt;NA&gt;
##    rs74315482    chr22  50625213      * |         &lt;NA&gt;              G                  A      &lt;NA&gt;
##           ...      ...       ...    ... .          ...            ...                ...       ...
##   rs199476369    chr22  50625936      * |         &lt;NA&gt;              C                  G      &lt;NA&gt;
##     rs2071421    chr22  50625988      * |         &lt;NA&gt;              T                  C      &lt;NA&gt;
##    rs74315475    chr22  50626033      * |         &lt;NA&gt;              T                  A      &lt;NA&gt;
##   rs398123419    chr22  50626052      * |         &lt;NA&gt;              C                  A      &lt;NA&gt;
##   rs398123418    chr22  50626057      * |         &lt;NA&gt;              G                  A      &lt;NA&gt;
##                    FILTER
##               &lt;character&gt;
##     rs6151429           .
##     rs6151428           .
##   rs774153480           .
##   rs199476388           .
##    rs74315482           .
##           ...         ...
##   rs199476369           .
##     rs2071421           .
##    rs74315475           .
##   rs398123419           .
##   rs398123418           .
##   -------
##   seqinfo: 25 sequences from hg19 genome; no seqlengths</code></pre>
</div>
</div>
<div id="sessioninfo" class="section level1">
<h1><span class="header-section-number">2</span> sessionInfo</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.5.0 RC (2018-04-16 r74624)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Sierra 10.12.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] BSgenome.Hsapiens.UCSC.hg19_1.4.0 BSgenome_1.47.5                  
##  [3] rtracklayer_1.39.13               BiocStyle_2.7.9                  
##  [5] VariantAnnotation_1.25.13         SummarizedExperiment_1.9.18      
##  [7] DelayedArray_0.5.35               BiocParallel_1.13.3              
##  [9] matrixStats_0.53.1                Rsamtools_1.31.3                 
## [11] Biostrings_2.47.12                XVector_0.19.9                   
## [13] GenomicFeatures_1.31.10           AnnotationDbi_1.41.6             
## [15] Biobase_2.39.2                    GenomicRanges_1.31.23            
## [17] GenomeInfoDb_1.15.5               IRanges_2.13.29                  
## [19] S4Vectors_0.17.43                 AnnotationHub_2.12.0             
## [21] BiocGenerics_0.25.3              
## 
## loaded via a namespace (and not attached):
##  [1] progress_1.1.2                xfun_0.1                      lattice_0.20-35              
##  [4] htmltools_0.3.6               yaml_2.1.19                   interactiveDisplayBase_1.18.0
##  [7] blob_1.1.1                    XML_3.98-1.11                 later_0.7.2                  
## [10] DBI_0.8                       bit64_0.9-7                   GenomeInfoDbData_1.1.0       
## [13] stringr_1.3.1                 zlibbioc_1.25.0               memoise_1.1.0                
## [16] evaluate_0.10.1               knitr_1.20                    biomaRt_2.35.13              
## [19] httpuv_1.4.3                  BiocInstaller_1.30.0          curl_3.2                     
## [22] Rcpp_0.12.17                  xtable_1.8-2                  promises_1.0.1               
## [25] backports_1.1.2               mime_0.5                      bit_1.1-12                   
## [28] digest_0.6.15                 stringi_1.2.2                 bookdown_0.7                 
## [31] shiny_1.0.5                   rprojroot_1.3-2               grid_3.5.0                   
## [34] tools_3.5.0                   bitops_1.0-6                  magrittr_1.5                 
## [37] RCurl_1.95-4.10               RSQLite_2.1.0                 pkgconfig_2.0.1              
## [40] Matrix_1.2-14                 prettyunits_1.0.2             assertthat_0.2.0             
## [43] rmarkdown_1.10                httr_1.3.1                    R6_2.2.2                     
## [46] GenomicAlignments_1.15.14     compiler_3.5.0</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
