<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>bioc_summarizedExperiment_exercises.utf8</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">seandavi(s12): Courses and Tutorials</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="setup.html">
    <span class="fa fa-cogs"></span>
     
    setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="ion ion-easel"></span>
     
    Slides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="motivation_for_R_slides.html">Motivation for using R</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
    Misc.
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="further_resources.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR">Source code for this site</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR/archive/master.zip">Download materials</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>




</div>


<div id="getting-started" class="section level1">
<h1>Getting started</h1>
<pre><code>## Loading required package: airway</code></pre>
<pre class="r"><code>library(SummarizedExperiment)
data(airway, package=&quot;airway&quot;)
airway</code></pre>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 64102 8 
## metadata(1): &#39;&#39;
## assays(1): counts
## rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99
## rowData names(0):
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
</div>
<div id="exercises" class="section level1">
<h1>Exercises</h1>
<ul>
<li>What is the <code>class()</code> of <code>airway</code>? What does the class tell you about the information stored about the features (genes) in the dataset?</li>
</ul>
<pre class="r"><code>class(airway)</code></pre>
<ul>
<li>How many features are in the data?</li>
</ul>
<pre class="r"><code>airway # it tells you
# but to get the result as a number:
nrow(airway)</code></pre>
<ul>
<li>How many samples are present?</li>
</ul>
<pre class="r"><code>airway # it tells you
# but to get the result as a number:
ncol(airway)</code></pre>
<ul>
<li>The assay information in the airway object contains the RNA-seq counts. Access the count data for the 56th gene.</li>
</ul>
<pre class="r"><code># check the names of the available assays
assayNames(airway)
# Both below are identical
assay(airway, &quot;counts&quot;)[56, ]
assays(airway)[[1]][56, ]</code></pre>
<ul>
<li>What are the <code>rownames()</code> of airway? What do these represent? What gene is represented by the first row of data?</li>
</ul>
<pre class="r"><code>head(rownames(airway))
# These are Ensembl gene identifiers
# ENSG00000000003 is the identifier of the &quot;TSPAN6&quot; gene</code></pre>
<ul>
<li>What information is available about the samples?</li>
</ul>
<pre class="r"><code>colData(airway)</code></pre>
<ul>
<li>How many samples were treated with dex? Untreated?</li>
</ul>
<pre class="r"><code>colData(airway) # and count
colData(airway)$dex # and count
table(colData(airway)$dex)</code></pre>
<ul>
<li>How many different cell types were used?</li>
</ul>
<pre class="r"><code>table(colData(airway)$cell)
length(unique(colData(airway)$cell))</code></pre>
<ul>
<li>What does <code>rowRanges(airway)</code> give you? What is the length of the object? What does each element of <code>rowRanges(airway)</code> contain?</li>
</ul>
<pre class="r"><code>rowRanges(airway)
length(rowRanges(airway))
rowRanges(airway)[[1]]</code></pre>
<ul>
<li>From the <a href="https://www.ncbi.nlm.nih.gov/pubmed/24926665">PubMed abstract</a>, we can see that the authors suggest that the DUSP1 gene is upregulated by Dexamethasone treatment. The <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=DUSP1&amp;keywords=DUSP1">GeneCards website</a> tells us that the Ensembl gene id for DUSP1 is ‘ENSG00000120129’. Get the count data for this gene.</li>
</ul>
<pre class="r"><code># get the assay data for DUSP1
dat = assay(airway, &quot;counts&quot;)[&quot;ENSG00000120129&quot;, ]</code></pre>
<ul>
<li>Make a plot of the gene counts in the previous exercise. Color the points based on the dexamethasone treatment.</li>
</ul>
<pre class="r"><code># get the assay data for DUSP1
plot(dat)</code></pre>
<p><img src="bioc_summarizedExperiment_exercises_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<pre class="r"><code>dextrt = colData(airway)$dex
plot(dat, col=dextrt)</code></pre>
<p><img src="bioc_summarizedExperiment_exercises_files/figure-html/unnamed-chunk-13-2.png" width="672" /></p>
</div>
<div id="sessioninfo" class="section level1">
<h1>sessionInfo()</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.5.0 RC (2018-04-16 r74624)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Sierra 10.12.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] airway_0.114.0                    knitr_1.20                       
##  [3] nycflights13_0.2.2                ggplot2_2.2.1                    
##  [5] BSgenome.Hsapiens.UCSC.hg19_1.4.0 BSgenome_1.47.5                  
##  [7] rtracklayer_1.39.13               BiocStyle_2.7.9                  
##  [9] VariantAnnotation_1.25.13         SummarizedExperiment_1.9.18      
## [11] DelayedArray_0.5.35               BiocParallel_1.13.3              
## [13] matrixStats_0.53.1                Rsamtools_1.31.3                 
## [15] Biostrings_2.47.12                XVector_0.19.9                   
## [17] GenomicFeatures_1.31.10           AnnotationDbi_1.41.6             
## [19] Biobase_2.39.2                    GenomicRanges_1.31.23            
## [21] GenomeInfoDb_1.15.5               IRanges_2.13.29                  
## [23] S4Vectors_0.17.43                 AnnotationHub_2.12.0             
## [25] BiocGenerics_0.25.3              
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.3.1                    bit64_0.9-7                  
##  [3] shiny_1.0.5                   assertthat_0.2.0             
##  [5] interactiveDisplayBase_1.18.0 blob_1.1.1                   
##  [7] GenomeInfoDbData_1.1.0        yaml_2.1.19                  
##  [9] progress_1.1.2                pillar_1.2.2                 
## [11] RSQLite_2.1.0                 backports_1.1.2              
## [13] lattice_0.20-35               digest_0.6.15                
## [15] promises_1.0.1                colorspace_1.3-2             
## [17] htmltools_0.3.6               httpuv_1.4.3                 
## [19] Matrix_1.2-14                 plyr_1.8.4                   
## [21] XML_3.98-1.11                 pkgconfig_2.0.1              
## [23] biomaRt_2.35.13               bookdown_0.7                 
## [25] zlibbioc_1.25.0               xtable_1.8-2                 
## [27] scales_0.5.0                  later_0.7.2                  
## [29] tibble_1.4.2                  lazyeval_0.2.1               
## [31] cli_1.0.0                     magrittr_1.5                 
## [33] crayon_1.3.4                  mime_0.5                     
## [35] memoise_1.1.0                 evaluate_0.10.1              
## [37] BiocInstaller_1.30.0          tools_3.5.0                  
## [39] prettyunits_1.0.2             stringr_1.3.1                
## [41] munsell_0.4.3                 compiler_3.5.0               
## [43] rlang_0.2.0                   grid_3.5.0                   
## [45] RCurl_1.95-4.10               bitops_1.0-6                 
## [47] labeling_0.3                  rmarkdown_1.9                
## [49] gtable_0.2.0                  DBI_0.8                      
## [51] curl_3.2                      R6_2.2.2                     
## [53] GenomicAlignments_1.15.14     utf8_1.1.3                   
## [55] bit_1.1-12                    rprojroot_1.3-2              
## [57] stringi_1.2.2                 Rcpp_0.12.16                 
## [59] xfun_0.1</code></pre>
</div>

    <script>
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
        ga('create', 'UA-93043521-1', 'auto');
        ga('send', 'pageview');

         
        var links = document.querySelectorAll('a');
        Array.prototype.map.call(links, function(item) {
            if (item.host != document.location.host) {
                item.addEventListener('click', function() {
                    var action = item.getAttribute('data-action') || 'follow';
                    ga('send', 'event', 'outbound', action, item.href);
                });
            }
        });
    </script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
