<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Control structures and functions</title>
  <meta name="description" content="Control structures and functions">
  <meta name="generator" content="bookdown 0.4.1 and GitBook 2.6.7">

  <meta property="og:title" content="Control structures and functions" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Control structures and functions" />
  
  
  

<meta name="author" content="Sean Davis">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="graphics-and-visualization.html">
<link rel="next" href="appendix-data-sets.html">
<script src="book_assets/jquery-2.2.3/jquery.min.js"></script>
<link href="book_assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Why R?</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#what-is-r"><i class="fa fa-check"></i><b>1.1</b> What is R?</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#why-use-r"><i class="fa fa-check"></i><b>1.2</b> Why use R?</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#why-not-use-r"><i class="fa fa-check"></i><b>1.3</b> Why not use R?</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#r-license-and-the-open-source-ideal"><i class="fa fa-check"></i><b>1.4</b> R License and the Open Source Ideal</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="r-mechanics.html"><a href="r-mechanics.html"><i class="fa fa-check"></i><b>2</b> R Mechanics</a><ul>
<li class="chapter" data-level="2.1" data-path="r-mechanics.html"><a href="r-mechanics.html#installing-r"><i class="fa fa-check"></i><b>2.1</b> Installing R</a><ul>
<li class="chapter" data-level="2.1.1" data-path="r-mechanics.html"><a href="r-mechanics.html#windows"><i class="fa fa-check"></i><b>2.1.1</b> Windows</a></li>
<li class="chapter" data-level="2.1.2" data-path="r-mechanics.html"><a href="r-mechanics.html#mac"><i class="fa fa-check"></i><b>2.1.2</b> Mac</a></li>
<li class="chapter" data-level="2.1.3" data-path="r-mechanics.html"><a href="r-mechanics.html#linux"><i class="fa fa-check"></i><b>2.1.3</b> Linux</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="r-mechanics.html"><a href="r-mechanics.html#starting-r"><i class="fa fa-check"></i><b>2.2</b> Starting R</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="first-steps.html"><a href="first-steps.html"><i class="fa fa-check"></i><b>3</b> First steps</a><ul>
<li class="chapter" data-level="3.1" data-path="first-steps.html"><a href="first-steps.html#interacting-with-r"><i class="fa fa-check"></i><b>3.1</b> Interacting with R</a></li>
<li class="chapter" data-level="3.2" data-path="first-steps.html"><a href="first-steps.html#rules-for-names-in-r"><i class="fa fa-check"></i><b>3.2</b> Rules for Names in R</a></li>
<li class="chapter" data-level="3.3" data-path="first-steps.html"><a href="first-steps.html#resources-for-getting-help"><i class="fa fa-check"></i><b>3.3</b> Resources for Getting Help</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html"><i class="fa fa-check"></i><b>4</b> Introduction to R data structures</a><ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a><ul>
<li class="chapter" data-level="4.1.1" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#creating-vectors"><i class="fa fa-check"></i><b>4.1.1</b> Creating vectors</a></li>
<li class="chapter" data-level="4.1.2" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#vector-operations"><i class="fa fa-check"></i><b>4.1.2</b> Vector Operations</a></li>
<li class="chapter" data-level="4.1.3" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#logical-vectors"><i class="fa fa-check"></i><b>4.1.3</b> Logical Vectors</a></li>
<li class="chapter" data-level="4.1.4" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#logical-operators"><i class="fa fa-check"></i><b>4.1.4</b> Logical Operators</a></li>
<li class="chapter" data-level="4.1.5" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>4.1.5</b> Indexing Vectors</a></li>
<li class="chapter" data-level="4.1.6" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#character-vectors-a.k.a.-strings"><i class="fa fa-check"></i><b>4.1.6</b> Character Vectors, A.K.A. Strings</a></li>
<li class="chapter" data-level="4.1.7" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#missing-values-aka-na"><i class="fa fa-check"></i><b>4.1.7</b> Missing Values, AKA “NA”</a></li>
<li class="chapter" data-level="4.1.8" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#factors"><i class="fa fa-check"></i><b>4.1.8</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#rectangular-data"><i class="fa fa-check"></i><b>4.2</b> Rectangular Data</a><ul>
<li class="chapter" data-level="4.2.1" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#matrices"><i class="fa fa-check"></i><b>4.2.1</b> Matrices</a></li>
<li class="chapter" data-level="4.2.2" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.2.2</b> Data Frames</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="introduction-to-r-data-structures.html"><a href="introduction-to-r-data-structures.html#lists-and-objects"><i class="fa fa-check"></i><b>4.3</b> Lists and Objects</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="plotting-and-graphics.html"><a href="plotting-and-graphics.html"><i class="fa fa-check"></i><b>5</b> Plotting and Graphics</a><ul>
<li class="chapter" data-level="5.1" data-path="plotting-and-graphics.html"><a href="plotting-and-graphics.html#base-r-graphics"><i class="fa fa-check"></i><b>5.1</b> Base R graphics</a></li>
<li class="chapter" data-level="5.2" data-path="plotting-and-graphics.html"><a href="plotting-and-graphics.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.2</b> Plotting with ggplot2</a><ul>
<li class="chapter" data-level="5.2.1" data-path="plotting-and-graphics.html"><a href="plotting-and-graphics.html#mtcars-data"><i class="fa fa-check"></i><b>5.2.1</b> mtcars data</a></li>
<li class="chapter" data-level="5.2.2" data-path="plotting-and-graphics.html"><a href="plotting-and-graphics.html#nyc-flight-data"><i class="fa fa-check"></i><b>5.2.2</b> NYC Flight data</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="plotting-and-graphics.html"><a href="plotting-and-graphics.html#graphics-devices-and-saving-plots"><i class="fa fa-check"></i><b>5.3</b> Graphics Devices and Saving Plots</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html"><i class="fa fa-check"></i><b>6</b> Control Structures, Looping, and Applying</a><ul>
<li class="chapter" data-level="6.1" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#control-structures-and-looping"><i class="fa fa-check"></i><b>6.1</b> Control Structures and Looping</a><ul>
<li class="chapter" data-level="6.1.1" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#control-structures-in-r"><i class="fa fa-check"></i><b>6.1.1</b> Control Structures in R</a></li>
<li class="chapter" data-level="6.1.2" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#control-structure-and-looping-examples"><i class="fa fa-check"></i><b>6.1.2</b> Control Structure and Looping Examples</a></li>
<li class="chapter" data-level="6.1.3" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#control-structure-and-looping-examples-1"><i class="fa fa-check"></i><b>6.1.3</b> Control Structure and Looping Examples</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#applying"><i class="fa fa-check"></i><b>6.2</b> Applying</a><ul>
<li class="chapter" data-level="6.2.1" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#why-does-r-have-apply-functions"><i class="fa fa-check"></i><b>6.2.1</b> Why Does R Have Apply Functions</a></li>
<li class="chapter" data-level="6.2.2" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#apply-function-exercise"><i class="fa fa-check"></i><b>6.2.2</b> Apply Function Exercise</a></li>
<li class="chapter" data-level="6.2.3" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#related-apply-functions"><i class="fa fa-check"></i><b>6.2.3</b> Related Apply Functions</a></li>
<li class="chapter" data-level="6.2.4" data-path="control-structures-looping-and-applying.html"><a href="control-structures-looping-and-applying.html#related-apply-function-examples"><i class="fa fa-check"></i><b>6.2.4</b> Related Apply Function Examples</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>7</b> Functions</a><ul>
<li class="chapter" data-level="7.0.1" data-path="functions.html"><a href="functions.html#function-overview"><i class="fa fa-check"></i><b>7.0.1</b> Function Overview</a></li>
<li class="chapter" data-level="7.0.2" data-path="functions.html"><a href="functions.html#example-functions"><i class="fa fa-check"></i><b>7.0.2</b> Example Functions</a></li>
<li class="chapter" data-level="7.0.3" data-path="functions.html"><a href="functions.html#further-reading"><i class="fa fa-check"></i><b>7.0.3</b> Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-input-and-exploration.html"><a href="data-input-and-exploration.html"><i class="fa fa-check"></i><b>8</b> Data Input and Exploration</a><ul>
<li class="chapter" data-level="8.1" data-path="data-input-and-exploration.html"><a href="data-input-and-exploration.html#behavioral-risk-factor-surveillance-system"><i class="fa fa-check"></i><b>8.1</b> Behavioral Risk Factor Surveillance System</a></li>
<li class="chapter" data-level="8.2" data-path="data-input-and-exploration.html"><a href="data-input-and-exploration.html#all-phenotypic-data"><i class="fa fa-check"></i><b>8.2</b> ALL Phenotypic Data</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="exploration-and-simple-univariate-measures.html"><a href="exploration-and-simple-univariate-measures.html"><i class="fa fa-check"></i><b>9</b> Exploration and simple univariate measures</a><ul>
<li class="chapter" data-level="9.1" data-path="exploration-and-simple-univariate-measures.html"><a href="exploration-and-simple-univariate-measures.html#clean-data"><i class="fa fa-check"></i><b>9.1</b> Clean data</a></li>
<li class="chapter" data-level="9.2" data-path="exploration-and-simple-univariate-measures.html"><a href="exploration-and-simple-univariate-measures.html#weight-in-1990-vs.2010-females"><i class="fa fa-check"></i><b>9.2</b> Weight in 1990 vs. 2010 Females</a></li>
<li class="chapter" data-level="9.3" data-path="exploration-and-simple-univariate-measures.html"><a href="exploration-and-simple-univariate-measures.html#weight-and-height-in-2010-males"><i class="fa fa-check"></i><b>9.3</b> Weight and height in 2010 Males</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="multivariate-analysis.html"><a href="multivariate-analysis.html"><i class="fa fa-check"></i><b>10</b> Multivariate analysis</a><ul>
<li class="chapter" data-level="10.1" data-path="multivariate-analysis.html"><a href="multivariate-analysis.html#input-and-setup"><i class="fa fa-check"></i><b>10.1</b> Input and setup</a></li>
<li class="chapter" data-level="10.2" data-path="multivariate-analysis.html"><a href="multivariate-analysis.html#cleaning"><i class="fa fa-check"></i><b>10.2</b> Cleaning</a></li>
<li class="chapter" data-level="10.3" data-path="multivariate-analysis.html"><a href="multivariate-analysis.html#unsupervised-machine-learning-multi-dimensional-scaling"><i class="fa fa-check"></i><b>10.3</b> Unsupervised machine learning – multi-dimensional scaling</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="using-r-in-real-life.html"><a href="using-r-in-real-life.html"><i class="fa fa-check"></i><b>11</b> Using <em>R</em> in real life</a><ul>
<li class="chapter" data-level="11.1" data-path="using-r-in-real-life.html"><a href="using-r-in-real-life.html#organizing-work"><i class="fa fa-check"></i><b>11.1</b> Organizing work</a></li>
<li class="chapter" data-level="11.2" data-path="using-r-in-real-life.html"><a href="using-r-in-real-life.html#r-packages"><i class="fa fa-check"></i><b>11.2</b> <em>R</em> Packages</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="graphics-and-visualization.html"><a href="graphics-and-visualization.html"><i class="fa fa-check"></i><b>12</b> Graphics and Visualization</a><ul>
<li class="chapter" data-level="12.1" data-path="graphics-and-visualization.html"><a href="graphics-and-visualization.html#base-r-graphics-1"><i class="fa fa-check"></i><b>12.1</b> Base <em>R</em> Graphics</a></li>
<li class="chapter" data-level="12.2" data-path="graphics-and-visualization.html"><a href="graphics-and-visualization.html#what-makes-for-a-good-graphical-display"><i class="fa fa-check"></i><b>12.2</b> What makes for a good graphical display?</a></li>
<li class="chapter" data-level="12.3" data-path="graphics-and-visualization.html"><a href="graphics-and-visualization.html#grammar-of-graphics-ggplot2"><i class="fa fa-check"></i><b>12.3</b> Grammar of Graphics: ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html"><i class="fa fa-check"></i><b>13</b> <em>Bioconductor</em> ‘infrastructure’ for sequence analysis</a><ul>
<li class="chapter" data-level="13.1" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#introduction"><i class="fa fa-check"></i><b>13.1</b> Introduction</a><ul>
<li class="chapter" data-level="13.1.1" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#classes-methods-and-packages"><i class="fa fa-check"></i><b>13.1.1</b> Classes, methods, and packages</a></li>
<li class="chapter" data-level="13.1.2" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#motivation"><i class="fa fa-check"></i><b>13.1.2</b> Motivation</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#core-packages"><i class="fa fa-check"></i><b>13.2</b> Core packages</a></li>
<li class="chapter" data-level="13.3" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#iranges-and-granges"><i class="fa fa-check"></i><b>13.3</b> <em>IRanges</em> and <em>GRanges</em></a><ul>
<li class="chapter" data-level="13.3.1" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#the-granges-and-grangeslist-classes"><i class="fa fa-check"></i><b>13.3.1</b> The <code>GRanges</code> and <code>GRangesList</code> classes</a></li>
<li class="chapter" data-level="13.3.2" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#algebra-of-genomic-ranges"><i class="fa fa-check"></i><b>13.3.2</b> Algebra of genomic ranges</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#biostrings-dna-or-amino-acid-sequences"><i class="fa fa-check"></i><b>13.4</b> <em>Biostrings</em> (DNA or amino acid sequences)</a></li>
<li class="chapter" data-level="13.5" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#other-formats-and-packages"><i class="fa fa-check"></i><b>13.5</b> Other formats and packages</a></li>
<li class="chapter" data-level="13.6" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#public-data-and-annotation-from-annotationhub"><i class="fa fa-check"></i><b>13.6</b> Public Data and Annotation from <em>AnnotationHub</em></a><ul>
<li class="chapter" data-level="13.6.1" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#roadmap-epigenomics-project"><i class="fa fa-check"></i><b>13.6.1</b> Roadmap Epigenomics Project</a></li>
<li class="chapter" data-level="13.6.2" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#ensembl-gtf-txdb-gene-models"><i class="fa fa-check"></i><b>13.6.2</b> Ensembl GTF TxDb gene models</a></li>
</ul></li>
<li class="chapter" data-level="13.7" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#epigenomics-case-study"><i class="fa fa-check"></i><b>13.7</b> Epigenomics case study</a><ul>
<li class="chapter" data-level="13.7.1" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#peak-density-across-the-genome"><i class="fa fa-check"></i><b>13.7.1</b> Peak density across the genome</a></li>
<li class="chapter" data-level="13.7.2" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#compare-peak-sets"><i class="fa fa-check"></i><b>13.7.2</b> Compare peak sets</a></li>
<li class="chapter" data-level="13.7.3" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#histone-profiles"><i class="fa fa-check"></i><b>13.7.3</b> Histone profiles</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="bioconductor-infrastructure-for-sequence-analysis.html"><a href="bioconductor-infrastructure-for-sequence-analysis.html#sessioninfo"><i class="fa fa-check"></i><b>13.8</b> <code>sessionInfo()</code></a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="appendix-data-sets.html"><a href="appendix-data-sets.html"><i class="fa fa-check"></i><b>A</b> Appendix – Data Sets</a></li>
<li class="chapter" data-level="B" data-path="appendix-swirl.html"><a href="appendix-swirl.html"><i class="fa fa-check"></i><b>B</b> Appendix – Swirl</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Control structures and functions</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="bioconductor-infrastructure-for-sequence-analysis" class="section level1">
<h1><span class="header-section-number">13</span> <em>Bioconductor</em> ‘infrastructure’ for sequence analysis</h1>
<p>We will need to install a few bioconductor packages for this section.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&#39;https://bioconductor.org/biocLite.R&#39;</span>)
<span class="kw">biocLite</span>(<span class="kw">c</span>(<span class="st">&quot;GenomicAlignments&quot;</span>,
           <span class="st">&quot;BSgenome.Hsapiens.UCSC.hg19&quot;</span>,
           <span class="st">&quot;AnnotationHub&quot;</span>,
           <span class="st">&quot;TxDb.Hsapiens.BioMart.igis&quot;</span>,
           <span class="st">&quot;rtracklayer&quot;</span>,
           <span class="st">&quot;readr&quot;</span>,
           <span class="st">&quot;ComplexHeatmap&quot;</span>,
           <span class="st">&quot;ggbio&quot;</span>))</code></pre></div>
<div id="introduction" class="section level2">
<h2><span class="header-section-number">13.1</span> Introduction</h2>
<div id="classes-methods-and-packages" class="section level3">
<h3><span class="header-section-number">13.1.1</span> Classes, methods, and packages</h3>
<p>This section focuses on classes, methods, and packages, with the goal being to learn to navigate the help system and interactive discovery facilities.</p>
</div>
<div id="motivation" class="section level3">
<h3><span class="header-section-number">13.1.2</span> Motivation</h3>
<p>Sequence analysis is specialized</p>
<ul>
<li>Large data needs to be processed in a memory- and time-efficient manner</li>
<li>Specific algorithms have been developed for the unique characteristics of sequence data</li>
</ul>
<p>Additional considerations</p>
<ul>
<li>Re-use of existing, tested code is easier to do and less error-prone than re-inventing the wheel.</li>
<li>Interoperability between packages is easier when the packages share similar data structures.</li>
</ul>
<p>Solution: use well-defined <em>classes</em> to represent complex data; <em>methods</em> operate on the classes to perform useful functions. Classes and methods are placed together and distributed as <em>packages</em> so that we can all benefit from the hard work and tested code of others.</p>
</div>
</div>
<div id="core-packages" class="section level2">
<h2><span class="header-section-number">13.2</span> Core packages</h2>
<pre>
                   VariantAnnotation
                           |
                           v
                    GenomicFeatures
                           |
                           v
                       BSgenome
                           |
                           v
                      rtracklayer
                           |
                           v
                    GenomicAlignments
                      |           |
                      v           v
     SummarizedExperiment   Rsamtools  ShortRead
                  |         |      |      |
                  v         v      v      v
                GenomicRanges     Biostrings
                        |          |
                        v          v
               GenomeInfoDb   (XVector)
                        |     |
                        v     v
                        IRanges
                           |
                           v 
                      (S4Vectors)
</pre>
</div>
<div id="iranges-and-granges" class="section level2">
<h2><span class="header-section-number">13.3</span> <em>IRanges</em> and <em>GRanges</em></h2>
<p>The <a href="http://bioconductor.org/packages/IRanges">IRanges</a> package defines an important class for specifying integer ranges, e.g.,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(IRanges)
ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>), <span class="dt">width=</span><span class="dv">5</span>)
ir</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        10        14         5
##   [2]        20        24         5
##   [3]        30        34         5</code></pre>
<p>There are many interesting operations to be performed on ranges, e.g, <code>flank()</code> identifies adjacent ranges</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">flank</span>(ir, <span class="dv">3</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         7         9         3
##   [2]        17        19         3
##   [3]        27        29         3</code></pre>
<p>The <code>IRanges</code> class is part of a class hierarchy. To see this, ask R for the class of <code>ir</code>, and for the class definition of the <code>IRanges</code> class</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(ir)</code></pre></div>
<pre><code>## [1] &quot;IRanges&quot;
## attr(,&quot;package&quot;)
## [1] &quot;IRanges&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getClass</span>(<span class="kw">class</span>(ir))</code></pre></div>
<pre><code>## Class &quot;IRanges&quot; [package &quot;IRanges&quot;]
## 
## Slots:
##                                                                                                 
## Name:              start             width             NAMES       elementType   elementMetadata
## Class:           integer           integer character_OR_NULL         character DataTable_OR_NULL
##                         
## Name:           metadata
## Class:              list
## 
## Extends: 
## Class &quot;Ranges&quot;, directly
## Class &quot;IntegerList&quot;, by class &quot;Ranges&quot;, distance 2
## Class &quot;RangesORmissing&quot;, by class &quot;Ranges&quot;, distance 2
## Class &quot;AtomicList&quot;, by class &quot;Ranges&quot;, distance 3
## Class &quot;List&quot;, by class &quot;Ranges&quot;, distance 4
## Class &quot;Vector&quot;, by class &quot;Ranges&quot;, distance 5
## Class &quot;Annotated&quot;, by class &quot;Ranges&quot;, distance 6
## 
## Known Subclasses: &quot;NormalIRanges&quot;, &quot;GroupingIRanges&quot;</code></pre>
<p>Notice that <code>IRanges</code> extends the <code>Ranges</code> class. Show</p>
<p>Now try entering <code>?flank</code> (if not using <em>RStudio</em>, enter <code>?&quot;flank,&lt;tab&gt;&quot;</code> where <code>&lt;tab&gt;</code> means to press the tab key to ask for tab completion). You can see that there are help pages for <code>flank</code> operating on several different classes. Select the completion</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">?<span class="st">&quot;flank,Ranges-method&quot;</span> </code></pre></div>
<p>and verify that you’re at the page that describes the method relevant to an <code>IRanges</code> instance. Explore other range-based operations.</p>
<p>The <a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a> package extends the notion of ranges to include features relevant to application of ranges in sequence analysis, particularly the ability to associate a range with a sequence name (e.g., chromosome) and a strand. Create a <code>GRanges</code> instance based on our <code>IRanges</code> instance, as follows</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicRanges)
gr &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>, <span class="st">&quot;chr1&quot;</span>, <span class="st">&quot;chr2&quot;</span>), ir, <span class="dt">strand=</span><span class="kw">c</span>(<span class="st">&quot;+&quot;</span>, <span class="st">&quot;-&quot;</span>, <span class="st">&quot;+&quot;</span>))
gr</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1  [10, 14]      +
##   [2]     chr1  [20, 24]      -
##   [3]     chr2  [30, 34]      +
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The notion of flanking sequence has a more nuanced meaning in biology. In particular we might expect that flanking sequence on the <code>+</code> strand would precede the range, but on the minus strand would follow it. Verify that <code>flank</code> applied to a <code>GRanges</code> object has this behavior.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">flank</span>(gr, <span class="dv">3</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1  [ 7,  9]      +
##   [2]     chr1  [25, 27]      -
##   [3]     chr2  [27, 29]      +
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Discover what classes <code>GRanges</code> extends, find the help page documenting the behavior of <code>flank</code> when applied to a <code>GRanges</code> object,</p>
<p>It seems like there might be a number of helpful methods available for working with genomic ranges; we can discover some of these from the command line, indicating that the methods should be on the current <code>search()</code> path</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">methods</span>(<span class="dt">class=</span><span class="st">&quot;GRanges&quot;</span>)</code></pre></div>
<pre><code>##   [1] !=                  [                   [&lt;-                 %in%               
##   [5] &lt;                   &lt;=                  ==                  &gt;                  
##   [9] &gt;=                  $                   $&lt;-                 aggregate          
##  [13] anyNA               append              as.character        as.complex         
##  [17] as.data.frame       as.env              as.factor           as.integer         
##  [21] as.list             as.logical          as.matrix           as.numeric         
##  [25] as.raw              BamViews            bamWhich&lt;-          blocks             
##  [29] browseGenome        by                  c                   chrom              
##  [33] chrom&lt;-             coerce              coerce&lt;-            countOverlaps      
##  [37] coverage            disjoin             disjointBins        distance           
##  [41] distanceToNearest   duplicated          elementMetadata     elementMetadata&lt;-  
##  [45] end                 end&lt;-               eval                expand             
##  [49] expand.grid         export              extractROWS         extractUpstreamSeqs
##  [53] findOverlaps        flank               follow              gaps               
##  [57] getPromoterSeq      granges             head                intersect          
##  [61] is.na               is.unsorted         isDisjoint          length             
##  [65] lengths             liftOver            mapFromAlignments   mapFromTranscripts 
##  [69] mapToAlignments     mapToTranscripts    match               mcols              
##  [73] mcols&lt;-             merge               metadata            metadata&lt;-         
##  [77] mstack              names               names&lt;-             narrow             
##  [81] nearest             NROW                Ops                 order              
##  [85] overlapsAny         parallelSlotNames   pcompare            pgap               
##  [89] pintersect          pmapFromAlignments  pmapFromTranscripts pmapToAlignments   
##  [93] pmapToTranscripts   precede             promoters           psetdiff           
##  [97] punion              range               ranges              ranges&lt;-           
## [101] rank                reduce              relist              relistToClass      
## [105] rename              rep                 rep.int             replaceROWS        
## [109] resize              restrict            rev                 ROWNAMES           
## [113] rowRanges&lt;-         ScanBamParam        ScanBcfParam        scanFa             
## [117] scanTabix           score               score&lt;-             selfmatch          
## [121] seqinfo             seqinfo&lt;-           seqlevelsInUse      seqnames           
## [125] seqnames&lt;-          setdiff             setequal            shift              
## [129] shiftApply          show                showAsCell          slidingWindows     
## [133] sort                split               split&lt;-             start              
## [137] start&lt;-             strand              strand&lt;-            subset             
## [141] subsetByOverlaps    summarizeOverlaps   summary             table              
## [145] tail                tapply              tile                transform          
## [149] trim                union               unique              update             
## [153] updateObject        values              values&lt;-            width              
## [157] width&lt;-             window              window&lt;-            with               
## [161] xtabs              
## see &#39;?methods&#39; for accessing help and source code</code></pre>
<p>Notice that the available <code>flank()</code> methods have been augmented by the methods defined in the <em>GenomicRanges</em> package, including those that are relevant (via inheritance) to the <em>GRanges</em> class.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">grep</span>(<span class="st">&quot;flank&quot;</span>, <span class="kw">methods</span>(<span class="dt">class=</span><span class="st">&quot;GRanges&quot;</span>), <span class="dt">value=</span><span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## [1] &quot;flank,GenomicRanges-method&quot;</code></pre>
<p>Verify that the help page documents the behavior we just observed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">?<span class="st">&quot;flank,GenomicRanges-method&quot;</span></code></pre></div>
<p>Use <code>help()</code> to list the help pages in the <code>GenomicRanges</code> package, and <code>vignettes()</code> to view and access available vignettes; these are also available in the Rstudio ‘Help’ tab.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">help</span>(<span class="dt">package=</span><span class="st">&quot;GenomicRanges&quot;</span>)
<span class="kw">vignette</span>(<span class="dt">package=</span><span class="st">&quot;GenomicRanges&quot;</span>)
<span class="kw">vignette</span>(<span class="dt">package=</span><span class="st">&quot;GenomicRanges&quot;</span>, <span class="st">&quot;GenomicRangesHOWTOs&quot;</span>)</code></pre></div>
<div id="the-granges-and-grangeslist-classes" class="section level3">
<h3><span class="header-section-number">13.3.1</span> The <code>GRanges</code> and <code>GRangesList</code> classes</h3>
<p>Aside: ‘TxDb’ packages provide an R representation of gene models</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb &lt;-<span class="st"> </span>TxDb.Hsapiens.UCSC.hg19.knownGene</code></pre></div>
<p><code>exons()</code>: <em>GRanges</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">exons</span>(txdb)</code></pre></div>
<pre><code>## GRanges object with 289969 ranges and 1 metadata column:
##                  seqnames         ranges strand |   exon_id
##                     &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1 [11874, 12227]      + |         1
##        [2]           chr1 [12595, 12721]      + |         2
##        [3]           chr1 [12613, 12721]      + |         3
##        [4]           chr1 [12646, 12697]      + |         4
##        [5]           chr1 [13221, 14409]      + |         5
##        ...            ...            ...    ... .       ...
##   [289965] chrUn_gl000241 [35706, 35859]      - |    289965
##   [289966] chrUn_gl000241 [36711, 36875]      - |    289966
##   [289967] chrUn_gl000243 [11501, 11530]      + |    289967
##   [289968] chrUn_gl000243 [13608, 13637]      + |    289968
##   [289969] chrUn_gl000247 [ 5787,  5816]      - |    289969
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p><code>exonsBy()</code>: <em>GRangesList</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">exonsBy</span>(txdb, <span class="st">&quot;tx&quot;</span>)</code></pre></div>
<pre><code>## GRangesList object of length 82960:
## $1 
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames         ranges strand |   exon_id   exon_name exon_rank
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
##   [1]     chr1 [11874, 12227]      + |         1        &lt;NA&gt;         1
##   [2]     chr1 [12613, 12721]      + |         3        &lt;NA&gt;         2
##   [3]     chr1 [13221, 14409]      + |         5        &lt;NA&gt;         3
## 
## $2 
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames         ranges strand | exon_id exon_name exon_rank
##   [1]     chr1 [11874, 12227]      + |       1      &lt;NA&gt;         1
##   [2]     chr1 [12595, 12721]      + |       2      &lt;NA&gt;         2
##   [3]     chr1 [13403, 14409]      + |       6      &lt;NA&gt;         3
## 
## $3 
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames         ranges strand | exon_id exon_name exon_rank
##   [1]     chr1 [11874, 12227]      + |       1      &lt;NA&gt;         1
##   [2]     chr1 [12646, 12697]      + |       4      &lt;NA&gt;         2
##   [3]     chr1 [13221, 14409]      + |       5      &lt;NA&gt;         3
## 
## ...
## &lt;82957 more elements&gt;
## -------
## seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p><em>GRanges</em> / <em>GRangesList</em> are incredibly useful</p>
<ul>
<li>Represent <strong>annotations</strong> – genes, variants, regulatory elements, copy number regions, …</li>
<li>Represent <strong>data</strong> – aligned reads, ChIP peaks, called variants, …</li>
</ul>
</div>
<div id="algebra-of-genomic-ranges" class="section level3">
<h3><span class="header-section-number">13.3.2</span> Algebra of genomic ranges</h3>
<p>Many biologically interesting questions represent operations on ranges</p>
<ul>
<li>Count overlaps between aligned reads and known genes – <code>GenomicRanges::summarizeOverlaps()</code></li>
<li>Genes nearest to regulatory regions – <code>GenomicRanges::nearest()</code>, [ChIPseeker][]</li>
<li>Called variants relevant to clinical phenotypes – <a href="http://bioconductor.org/packages/VariantFiltering">VariantFiltering</a></li>
</ul>
<p><em>GRanges</em> Algebra</p>
<ul>
<li>Intra-range methods
<ul>
<li>Independent of other ranges in the same object</li>
<li>GRanges variants strand-aware</li>
<li><code>shift()</code>, <code>narrow()</code>, <code>flank()</code>, <code>promoters()</code>, <code>resize()</code>, <code>restrict()</code>, <code>trim()</code></li>
<li>See <code>?&quot;intra-range-methods&quot;</code></li>
</ul></li>
<li>Inter-range methods
<ul>
<li>Depends on other ranges in the same object</li>
<li><code>range()</code>, <code>reduce()</code>, <code>gaps()</code>, <code>disjoin()</code></li>
<li><code>coverage()</code> (!)</li>
<li>see <code>?&quot;inter-range-methods&quot;</code></li>
</ul></li>
<li>Between-range methods
<ul>
<li>Functions of two (or more) range objects</li>
<li><code>findOverlaps()</code>, <code>countOverlaps()</code>, …, <code>%over%</code>, <code>%within%</code>, <code>%outside%</code>; <code>union()</code>, <code>intersect()</code>, <code>setdiff()</code>, <code>punion()</code>, <code>pintersect()</code>, <code>psetdiff()</code></li>
</ul></li>
</ul>
<div class="figure">
<img src="images/RangeOperations.png" alt="Alt Ranges Algebra" />
<p class="caption">Alt Ranges Algebra</p>
</div>
</div>
</div>
<div id="biostrings-dna-or-amino-acid-sequences" class="section level2">
<h2><span class="header-section-number">13.4</span> <em>Biostrings</em> (DNA or amino acid sequences)</h2>
<p>Classes</p>
<ul>
<li>XString, XStringSet, e.g., DNAString (genomes), DNAStringSet (reads)</li>
</ul>
<p>Methods –</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/bioc/vignettes/Biostrings/inst/doc/BiostringsQuickOverview.pdf">Cheat sheat</a></li>
<li>Manipulation, e.g., <code>reverseComplement()</code></li>
<li>Summary, e.g., <code>letterFrequency()</code></li>
<li>Matching, e.g., <code>matchPDict()</code>, <code>matchPWM()</code></li>
</ul>
<p>Related packages</p>
<ul>
<li><a href="http://bioconductor.org/packages/BSgenome">BSgenome</a></li>
<li>Whole-genome representations</li>
<li>Model and custom</li>
<li><a href="http://bioconductor.org/packages/ShortRead">ShortRead</a></li>
<li>FASTQ files</li>
</ul>
<p>Example</p>
<ul>
<li><p>Whole-genome sequences are distrubuted by ENSEMBL, NCBI, and others as FASTA files; model organism whole genome sequences are packaged into more user-friendly <code>BSgenome</code> packages. The following calculates GC content across chr14.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)
chr14_range =<span class="st"> </span><span class="kw">GRanges</span>(<span class="st">&quot;chr14&quot;</span>, <span class="kw">IRanges</span>(<span class="dv">1</span>, <span class="kw">seqlengths</span>(Hsapiens)[<span class="st">&quot;chr14&quot;</span>]))
chr14_dna &lt;-<span class="st"> </span><span class="kw">getSeq</span>(Hsapiens, chr14_range)
<span class="kw">letterFrequency</span>(chr14_dna, <span class="st">&quot;GC&quot;</span>, <span class="dt">as.prob=</span><span class="ot">TRUE</span>)</code></pre></div>
<pre><code>##           G|C
## [1,] 0.336276</code></pre></li>
</ul>
</div>
<div id="other-formats-and-packages" class="section level2">
<h2><span class="header-section-number">13.5</span> Other formats and packages</h2>
<div class="figure">
<img src="images/FilesToPackages.png" alt="Alt Files and the Bioconductor packages that input them" />
<p class="caption">Alt Files and the Bioconductor packages that input them</p>
</div>
<p>Acknowledgements</p>
<ul>
<li><p>Core (Seattle): Sonali Arora, Marc Carlson, Nate Hayden, Jim Hester, Valerie Obenchain, Hervé Pagès, Paul Shannon, Dan Tenenbaum.</p></li>
<li><p>The research reported in this presentation was supported by the National Cancer Institute and the National Human Genome Research Institute of the National Institutes of Health under Award numbers U24CA180996 and U41HG004059, and the National Science Foundation under Award number 1247813. The content is solely the responsibility of the authors and does not necessarily represent the official views of the National Institutes of Health or the National Science Foundation.</p></li>
</ul>
</div>
<div id="public-data-and-annotation-from-annotationhub" class="section level2">
<h2><span class="header-section-number">13.6</span> Public Data and Annotation from <em>AnnotationHub</em></h2>
<div id="roadmap-epigenomics-project" class="section level3">
<h3><span class="header-section-number">13.6.1</span> Roadmap Epigenomics Project</h3>
<p>All Roadmap Epigenomics files are hosted <a href="http://egg2.wustl.edu/roadmap/data/byFileType/">here</a>. If one had to download these files on their own, one would navigate through the web interface to find useful files, then use something like the following <em>R</em> code.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">url &lt;-<span class="st"> &quot;http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/broadPeak/E001-H3K4me1.broadPeak.gz&quot;</span>
filename &lt;-<span class="st">  </span><span class="kw">basename</span>(url)
<span class="kw">download.file</span>(url, <span class="dt">destfile=</span>filename)
<span class="cf">if</span> (<span class="kw">file.exists</span>(filename))
   data &lt;-<span class="st"> </span><span class="kw">import</span>(filename, <span class="dt">format=</span><span class="st">&quot;bed&quot;</span>)</code></pre></div>
<p>This would have to be repeated for all files, and the onus would lie on the user to identify, download, import, and manage the local disk location of these files.</p>
<p>AnnotationHub reduces this task to just a few lines of <em>R</em> code</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(AnnotationHub)</code></pre></div>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: methods</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap,
##     parApply, parCapply, parLapply, parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, cbind, colMeans, colnames, colSums, do.call,
##     duplicated, eval, evalq, Filter, Find, get, grep, grepl, intersect, is.unsorted,
##     lapply, lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int, pmin,
##     pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames, rowSums, sapply,
##     setdiff, sort, table, tapply, union, unique, unsplit, which, which.max, which.min</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ah =<span class="st"> </span><span class="kw">AnnotationHub</span>()</code></pre></div>
<pre><code>## Warning: database may not be current
##   database: &#39;/Users/sdavis2//.AnnotationHub/annotationhub.sqlite3&#39;
##   reason: do not know how to convert &#39;content(GET(url, proxy))&#39; to class &quot;POSIXct&quot;</code></pre>
<pre><code>## snapshotDate(): 2017-04-24</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">epiFiles &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="st">&quot;EpigenomeRoadMap&quot;</span>)</code></pre></div>
<p>A look at the value returned by <code>epiFiles</code> shows us that 18248 roadmap resources are available via AnnotationHub. Additional information about the files is also available, e.g., where the files came from (dataprovider), genome, species, sourceurl, sourcetypes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">epiFiles</code></pre></div>
<pre><code>## AnnotationHub with 18248 records
## # snapshotDate(): 2017-04-24 
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: BigWigFile, GRanges, data.frame
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH28856&quot;]]&#39; 
## 
##             title                                 
##   AH28856 | E001-H3K4me1.broadPeak.gz             
##   AH28857 | E001-H3K4me3.broadPeak.gz             
##   AH28858 | E001-H3K9ac.broadPeak.gz              
##   AH28859 | E001-H3K9me3.broadPeak.gz             
##   AH28860 | E001-H3K27me3.broadPeak.gz            
##   ...       ...                                   
##   AH49540 | E058_mCRF_FractionalMethylation.bigwig
##   AH49541 | E059_mCRF_FractionalMethylation.bigwig
##   AH49542 | E061_mCRF_FractionalMethylation.bigwig
##   AH49543 | E081_mCRF_FractionalMethylation.bigwig
##   AH49544 | E082_mCRF_FractionalMethylation.bigwig</code></pre>
<p>A good sanity check to ensure that we have files only from the Roadmap Epigenomics project is to check that all the files in the returned smaller hub object come from <em>Homo sapiens</em> and the hg19 genome</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">unique</span>(epiFiles<span class="op">$</span>species)</code></pre></div>
<pre><code>## [1] &quot;Homo sapiens&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">unique</span>(epiFiles<span class="op">$</span>genome)</code></pre></div>
<pre><code>## [1] &quot;hg19&quot;</code></pre>
<p>Broadly, one can get an idea of the different files from this project looking at the sourcetype</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(epiFiles<span class="op">$</span>sourcetype)</code></pre></div>
<pre><code>## 
##    BED BigWig    GTF    tab    Zip 
##   8298   9932      3      1     14</code></pre>
<p>To get a more descriptive idea of these different files one can use:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">sort</span>(<span class="kw">table</span>(epiFiles<span class="op">$</span>description), <span class="dt">decreasing=</span><span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## 
## Bigwig File containing -log10(p-value) signal tracks from EpigenomeRoadMap Project 
##                                                                               6881 
## Bigwig File containing fold enrichment signal tracks from EpigenomeRoadMap Project 
##                                                                               2947 
##    Narrow ChIP-seq peaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                               2894 
##     Broad ChIP-seq peaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                               2534 
##    Gapped ChIP-seq peaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                               2534 
##        Narrow DNasePeaks for consolidated epigenomes from EpigenomeRoadMap Project 
##                                                                                131</code></pre>
<p>The ‘metadata’ provided by the Roadmap Epigenomics Project is also available. Note that the information displayed about a hub with a single resource is quite different from the information displayed when the hub references more than one resource.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metadata.tab &lt;-<span class="st"> </span><span class="kw">query</span>(ah , <span class="kw">c</span>(<span class="st">&quot;EpigenomeRoadMap&quot;</span>, <span class="st">&quot;Metadata&quot;</span>))
metadata.tab</code></pre></div>
<pre><code>## AnnotationHub with 1 record
## # snapshotDate(): 2017-04-24 
## # names(): AH41830
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: data.frame
## # $title: EID_metadata.tab
## # $description: Metadata for EpigenomeRoadMap Project
## # $taxonomyid: 9606
## # $genome: hg19
## # $sourcetype: tab
## # $sourceurl: http://egg2.wustl.edu/roadmap/data/byFileType/metadata/EID_metadata.tab
## # $sourcelastmodifieddate: 2015-02-15
## # $sourcesize: 18035
## # $tags: c(&quot;EpigenomeRoadMap&quot;, &quot;Metadata&quot;) 
## # retrieve record with &#39;object[[&quot;AH41830&quot;]]&#39;</code></pre>
<p>So far we have been exploring information about resources, without downloading the resource to a local cache and importing it into R. One can retrieve the resource using <code>[[</code> as indicated at the end of the show method</p>
<pre><code>## loading from cache &#39;/Users/sdavis2//.AnnotationHub/47270&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metadata.tab &lt;-<span class="st"> </span>ah[[<span class="st">&quot;AH41830&quot;</span>]]</code></pre></div>
<pre><code>## loading from cache &#39;/Users/sdavis2//.AnnotationHub/47270&#39;</code></pre>
<p>The metadata.tab file is returned as a <em>data.frame</em>. The first 6 rows of the first 5 columns are shown here:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metadata.tab[<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<pre><code>##    EID    GROUP   COLOR          MNEMONIC                                   STD_NAME
## 1 E001      ESC #924965            ESC.I3                                ES-I3 Cells
## 2 E002      ESC #924965           ESC.WA7                               ES-WA7 Cells
## 3 E003      ESC #924965            ESC.H1                                   H1 Cells
## 4 E004 ES-deriv #4178AE ESDR.H1.BMP4.MESO H1 BMP4 Derived Mesendoderm Cultured Cells
## 5 E005 ES-deriv #4178AE ESDR.H1.BMP4.TROP H1 BMP4 Derived Trophoblast Cultured Cells
## 6 E006 ES-deriv #4178AE       ESDR.H1.MSC          H1 Derived Mesenchymal Stem Cells</code></pre>
<p>One can keep constructing different queries using multiple arguments to trim down these 18248 to get the files one wants. For example, to get the ChIP-Seq files for consolidated epigenomes, one could use</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bpChipEpi &lt;-<span class="st"> </span><span class="kw">query</span>(ah , <span class="kw">c</span>(<span class="st">&quot;EpigenomeRoadMap&quot;</span>, <span class="st">&quot;broadPeak&quot;</span>, <span class="st">&quot;chip&quot;</span>, <span class="st">&quot;consolidated&quot;</span>))</code></pre></div>
<p>To get all the bigWig signal files, one can query the hub using</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">allBigWigFiles &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="kw">c</span>(<span class="st">&quot;EpigenomeRoadMap&quot;</span>, <span class="st">&quot;BigWig&quot;</span>))</code></pre></div>
<p>To access the 15 state chromatin segmentations, one can use</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">seg &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="kw">c</span>(<span class="st">&quot;EpigenomeRoadMap&quot;</span>, <span class="st">&quot;segmentations&quot;</span>))</code></pre></div>
<p>If one is interested in getting all the files related to one sample</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">E126 &lt;-<span class="st"> </span><span class="kw">query</span>(ah , <span class="kw">c</span>(<span class="st">&quot;EpigenomeRoadMap&quot;</span>, <span class="st">&quot;E126&quot;</span>, <span class="st">&quot;H3K4ME2&quot;</span>))
E126</code></pre></div>
<pre><code>## AnnotationHub with 6 records
## # snapshotDate(): 2017-04-24 
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: BigWigFile, GRanges
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH29817&quot;]]&#39; 
## 
##             title                                  
##   AH29817 | E126-H3K4me2.broadPeak.gz              
##   AH30868 | E126-H3K4me2.narrowPeak.gz             
##   AH31801 | E126-H3K4me2.gappedPeak.gz             
##   AH32990 | E126-H3K4me2.fc.signal.bigwig          
##   AH34022 | E126-H3K4me2.pval.signal.bigwig        
##   AH40177 | E126-H3K4me2.imputed.pval.signal.bigwig</code></pre>
<p>Hub resources can also be selected using <code>$</code>, <code>subset()</code>, and <code>display()</code>; see the main <a href="AnnotationHub.html"><em>AnnotationHub</em> vignette</a> for additional detail.</p>
<p>Hub resources are imported as the appropriate <em>Bioconductor</em> object for use in further analysis. For example, peak files are returned as <em>GRanges</em> objects.</p>
<pre><code>## loading from cache &#39;/Users/sdavis2//.AnnotationHub/35257&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">peaks &lt;-<span class="st"> </span>E126[[<span class="st">&#39;AH29817&#39;</span>]]</code></pre></div>
<pre><code>## loading from cache &#39;/Users/sdavis2//.AnnotationHub/35257&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqinfo</span>(peaks)</code></pre></div>
<pre><code>## Seqinfo object with 93 sequences (1 circular) from hg19 genome:
##   seqnames       seqlengths isCircular genome
##   chr1            249250621      FALSE   hg19
##   chr2            243199373      FALSE   hg19
##   chr3            198022430      FALSE   hg19
##   chr4            191154276      FALSE   hg19
##   chr5            180915260      FALSE   hg19
##   ...                   ...        ...    ...
##   chrUn_gl000245      36651      FALSE   hg19
##   chrUn_gl000246      38154      FALSE   hg19
##   chrUn_gl000247      36422      FALSE   hg19
##   chrUn_gl000248      39786      FALSE   hg19
##   chrUn_gl000249      38502      FALSE   hg19</code></pre>
<p>BigWig files are returned as <em>BigWigFile</em> objects. A <em>BigWigFile</em> is a reference to a file on disk; the data in the file can be read in using <code>rtracklayer::import()</code>, perhaps querying these large files for particular genomic regions of interest as described on the help page <code>?import.bw</code>.</p>
<p>Each record inside AnnotationHub is associated with a unique identifier. Most <em>GRanges</em> objects returned by AnnotationHub contain the unique AnnotationHub identifier of the resource from which the <em>GRanges</em> is derived. This can come handy when working with the <em>GRanges</em> object for a while, and additional information about the object (e.g., the name of the file in the cache, or the original sourceurl for the data underlying the resource) that is being worked with.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">metadata</span>(peaks)</code></pre></div>
<pre><code>## $AnnotationHubName
## [1] &quot;AH29817&quot;
## 
## $`File Name`
## [1] &quot;E126-H3K4me2.broadPeak.gz&quot;
## 
## $`Data Source`
## [1] &quot;http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/broadPeak/E126-H3K4me2.broadPeak.gz&quot;
## 
## $Provider
## [1] &quot;BroadInstitute&quot;
## 
## $Organism
## [1] &quot;Homo sapiens&quot;
## 
## $`Taxonomy ID`
## [1] 9606</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ah[<span class="kw">metadata</span>(peaks)<span class="op">$</span>AnnotationHubName]<span class="op">$</span>sourceurl</code></pre></div>
<pre><code>## [1] &quot;http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/broadPeak/E126-H3K4me2.broadPeak.gz&quot;</code></pre>
</div>
<div id="ensembl-gtf-txdb-gene-models" class="section level3">
<h3><span class="header-section-number">13.6.2</span> Ensembl GTF TxDb gene models</h3>
<p><em>Bioconductor</em> represents gene models using ‘transcript’ databases. These are available via packages such as <em>TxDb.Hsapiens.UCSC.hg38.knownGene</em> or can be constructed using functions such as <code>GenomicFeatures::makeTxDbFromBiomart()</code>.</p>
<p><em>AnnotationHub</em> provides an easy way to work with gene models published by Ensembl. We are going to be working with human data that are mapped to GRCh37 (hg19). We have a choice of providers (UCSC, NCBI, Ensembl), but we will just choose Ensembl here to minimize the amount of ID changing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">query</span>(ah,<span class="kw">c</span>(<span class="st">&#39;gtf&#39;</span>,<span class="st">&#39;ensembl&#39;</span>,<span class="st">&#39;sapiens&#39;</span>,<span class="st">&#39;GRCh37&#39;</span>))</code></pre></div>
<pre><code>## AnnotationHub with 7 records
## # snapshotDate(): 2017-04-24 
## # $dataprovider: Ensembl
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH7558&quot;]]&#39; 
## 
##             title                     
##   AH7558  | Homo_sapiens.GRCh37.70.gtf
##   AH7619  | Homo_sapiens.GRCh37.69.gtf
##   AH7666  | Homo_sapiens.GRCh37.71.gtf
##   AH7726  | Homo_sapiens.GRCh37.72.gtf
##   AH7790  | Homo_sapiens.GRCh37.73.gtf
##   AH8753  | Homo_sapiens.GRCh37.74.gtf
##   AH10684 | Homo_sapiens.GRCh37.75.gtf</code></pre>
<p>We see that there is a GTF file describing gene models. The GTF file is imported as a <em>GRanges</em> instance, but with a number of columns that allow the <em>Granges</em> object to encode the complexity of exons, CDS, and utrs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gtf &lt;-<span class="st"> </span>ah[[<span class="st">&#39;AH10684&#39;</span>]]</code></pre></div>
<pre><code>## loading from cache &#39;/Users/sdavis2//.AnnotationHub/10684&#39;</code></pre>
<pre><code>## using guess work to populate seqinfo</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(gtf, <span class="dv">3</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 16 metadata columns:
##       seqnames         ranges strand |               source       type     score     phase
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; |             &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##   [1]        1 [11869, 14412]      + |           pseudogene       gene      &lt;NA&gt;      &lt;NA&gt;
##   [2]        1 [11869, 14409]      + | processed_transcript transcript      &lt;NA&gt;      &lt;NA&gt;
##   [3]        1 [11869, 12227]      + | processed_transcript       exon      &lt;NA&gt;      &lt;NA&gt;
##               gene_id   gene_name    gene_source gene_biotype   transcript_id transcript_name
##           &lt;character&gt; &lt;character&gt;    &lt;character&gt;  &lt;character&gt;     &lt;character&gt;     &lt;character&gt;
##   [1] ENSG00000223972     DDX11L1 ensembl_havana   pseudogene            &lt;NA&gt;            &lt;NA&gt;
##   [2] ENSG00000223972     DDX11L1 ensembl_havana   pseudogene ENST00000456328     DDX11L1-002
##   [3] ENSG00000223972     DDX11L1 ensembl_havana   pseudogene ENST00000456328     DDX11L1-002
##       transcript_source exon_number         exon_id         tag     ccds_id  protein_id
##             &lt;character&gt;   &lt;numeric&gt;     &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
##   [1]              &lt;NA&gt;        &lt;NA&gt;            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   [2]            havana        &lt;NA&gt;            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   [3]            havana           1 ENSE00002234944        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome</code></pre>
<p>It is trivial to make a transcript database, or <em>TxDb</em>, instance of these data. We will use this <em>TxDb</em> to quickly access gene annotation in genome context.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicFeatures)         <span class="co"># for makeTxDbFromGRanges</span>
txdb &lt;-<span class="st"> </span><span class="kw">makeTxDbFromGRanges</span>(gtf)</code></pre></div>
</div>
</div>
<div id="epigenomics-case-study" class="section level2">
<h2><span class="header-section-number">13.7</span> Epigenomics case study</h2>
<p>We are going to use data from the <a href="https://www.genome.gov/10005107/">ENCODE project</a> as a playground for using the Bioconductor ranges infrastructure. To keep things manageable, we will focus on only the K562 leukemia cell line and on histone marks. To keep the data smallish, but without loss of generality, we will use data that have been processed using a peak-caller and summarized as regions of statistically significant regions.</p>
<p>Bioconductor has processed all of the ENCODE data into Bioconductor data structures (as well as thousands of other datasets) and then stored and made available via the <em>AnnotationHub</em> system we saw above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">query</span>(ah,<span class="kw">c</span>(<span class="st">&#39;K562&#39;</span>,<span class="st">&#39;h3k&#39;</span>,<span class="st">&#39;E123&#39;</span>,<span class="st">&#39;narrow&#39;</span>))</code></pre></div>
<pre><code>## AnnotationHub with 10 records
## # snapshotDate(): 2017-04-24 
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH30830&quot;]]&#39; 
## 
##             title                      
##   AH30830 | E123-H3K4me1.narrowPeak.gz 
##   AH30831 | E123-H3K4me2.narrowPeak.gz 
##   AH30832 | E123-H3K4me3.narrowPeak.gz 
##   AH30833 | E123-H3K9ac.narrowPeak.gz  
##   AH30834 | E123-H3K9me1.narrowPeak.gz 
##   AH30835 | E123-H3K9me3.narrowPeak.gz 
##   AH30836 | E123-H3K27ac.narrowPeak.gz 
##   AH30837 | E123-H3K27me3.narrowPeak.gz
##   AH30838 | E123-H3K36me3.narrowPeak.gz
##   AH30839 | E123-H3K79me2.narrowPeak.gz</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">histoneSets =<span class="st"> </span><span class="kw">names</span>(<span class="kw">query</span>(ah,<span class="kw">c</span>(<span class="st">&#39;K562&#39;</span>,<span class="st">&#39;E123&#39;</span>,<span class="st">&#39;h3k&#39;</span>,<span class="st">&#39;narrow&#39;</span>)))</code></pre></div>
<p>The <code>histoneSets</code> variable now contains the <em>names</em> <em>AnnotationHub</em> datasets from histone modification experiments done in K562 cells and summarized using a bioinformatics protocol tunes to produce narrow peaks. However, the data still need to be downloaded and then loaded into R memory.</p>
<p><em>To load your own peaks into R, check out the <code>rtracklayer::import()</code> function.</em></p>
<div id="peak-density-across-the-genome" class="section level3">
<h3><span class="header-section-number">13.7.1</span> Peak density across the genome</h3>
<p>Start (arbitrarily) with the first <code>histoneSet</code> by loading the dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">histonePeaks =<span class="st"> </span>ah[[histoneSets[<span class="dv">1</span>]]]</code></pre></div>
<pre><code>## loading from cache &#39;/Users/sdavis2//.AnnotationHub/36270&#39;</code></pre>
<p>The first time the code above is executed, the data will be <em>both</em> downloaded <em>and</em> loaded into the variable <code>histonePeaks</code>. The <em>next</em> time that this line is run, a “cached” version of the data will be used, so no download needs to occur the second time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">si =<span class="st"> </span><span class="kw">seqinfo</span>(histonePeaks)
si</code></pre></div>
<pre><code>## Seqinfo object with 93 sequences (1 circular) from hg19 genome:
##   seqnames       seqlengths isCircular genome
##   chr1            249250621      FALSE   hg19
##   chr2            243199373      FALSE   hg19
##   chr3            198022430      FALSE   hg19
##   chr4            191154276      FALSE   hg19
##   chr5            180915260      FALSE   hg19
##   ...                   ...        ...    ...
##   chrUn_gl000245      36651      FALSE   hg19
##   chrUn_gl000246      38154      FALSE   hg19
##   chrUn_gl000247      36422      FALSE   hg19
##   chrUn_gl000248      39786      FALSE   hg19
##   chrUn_gl000249      38502      FALSE   hg19</code></pre>
<p>The next block of code looks long, but the main <em>purpose</em> of the code is to produce 100kb windows across the genomic so that we can count the number of H3K4me1 peaks in each 100kb window. Most of the lines are do deal with small issues of keeping chromosome names and lengths straight when transforming our data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fullchroms =<span class="st"> </span><span class="kw">as</span>(si,<span class="st">&#39;GRanges&#39;</span>)
<span class="kw">seqinfo</span>(fullchroms)</code></pre></div>
<pre><code>## Seqinfo object with 93 sequences (1 circular) from hg19 genome:
##   seqnames       seqlengths isCircular genome
##   chr1            249250621      FALSE   hg19
##   chr2            243199373      FALSE   hg19
##   chr3            198022430      FALSE   hg19
##   chr4            191154276      FALSE   hg19
##   chr5            180915260      FALSE   hg19
##   ...                   ...        ...    ...
##   chrUn_gl000245      36651      FALSE   hg19
##   chrUn_gl000246      38154      FALSE   hg19
##   chrUn_gl000247      36422      FALSE   hg19
##   chrUn_gl000248      39786      FALSE   hg19
##   chrUn_gl000249      38502      FALSE   hg19</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 100kb bins</span>
<span class="co"># This is the work--using tile()</span>
tile0.1mb =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">tile</span>(fullchroms,<span class="dt">width=</span><span class="fl">1e5</span>))
<span class="kw">seqinfo</span>(tile0.1mb) =<span class="st"> </span>si
tile0.1mb =<span class="st"> </span><span class="kw">keepStandardChromosomes</span>(tile0.1mb,<span class="dt">pruning.mode =</span> <span class="st">&#39;coarse&#39;</span>)
histonePeaks =<span class="st"> </span><span class="kw">keepStandardChromosomes</span>(histonePeaks,<span class="dt">pruning.mode =</span> <span class="st">&#39;coarse&#39;</span>)
<span class="kw">seqlevelsStyle</span>(tile0.1mb) =<span class="st"> </span><span class="kw">seqlevelsStyle</span>(histonePeaks)
histonePeaks=<span class="kw">dropSeqlevels</span>(histonePeaks,<span class="st">&#39;chrM&#39;</span>)
<span class="kw">genome</span>(tile0.1mb) =<span class="st"> &#39;hg19&#39;</span></code></pre></div>
<p>Counting peaks in each 100kb window is <em>VERY</em> fast using the <code>counterOverlaps()</code> method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tile0.1mb<span class="op">$</span>peakCount =<span class="st"> </span><span class="kw">countOverlaps</span>(tile0.1mb,histonePeaks)</code></pre></div>
<p>While the following plot is not as informative as it could be, sometimes using a circular layout is a good way to show an overview of full genome datasets. Data on the scale of 100kb is probably still too fine-level for such a plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggbio)
<span class="kw">ggbio</span>() <span class="op">+</span><span class="st"> </span><span class="kw">circle</span>(tile0.1mb,<span class="dt">geom=</span><span class="st">&#39;point&#39;</span>, <span class="kw">aes</span>(<span class="dt">y =</span> peakCount),
                     <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">alpha=</span><span class="fl">0.25</span>, <span class="dt">size=</span><span class="fl">0.25</span>, <span class="dt">grid=</span><span class="ot">TRUE</span>,
                     <span class="dt">radius=</span><span class="dv">60</span>, <span class="dt">trackWidth=</span><span class="dv">30</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">circle</span>(tile0.1mb, <span class="dt">geom =</span> <span class="st">&quot;scale&quot;</span>, <span class="dt">size =</span> <span class="dv">2</span>,<span class="dt">radius=</span><span class="dv">90</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">circle</span>(tile0.1mb, <span class="dt">geom=</span><span class="st">&#39;ideo&#39;</span>,<span class="dt">fill=</span><span class="st">&#39;gray70&#39;</span>)</code></pre></div>
<p><img src="F_epigenomic_ranges_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p>
<p>Getting the <code>GC</code> content of the peaks is also straightforward. When you have two peaksets</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dnaInPeaks =<span class="st"> </span><span class="kw">getSeq</span>(BSgenome.Hsapiens.UCSC.hg19,histonePeaks)
percentGC =<span class="st"> </span><span class="kw">letterFrequency</span>(dnaInPeaks,<span class="st">&#39;GC&#39;</span>,<span class="dt">as.prob=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>To get a comparison set, just shift all regions to the left by 5kb (for example).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shiftedPeaks =<span class="st"> </span><span class="kw">shift</span>(histonePeaks,<span class="op">-</span><span class="dv">5000</span>)
<span class="kw">head</span>(<span class="kw">ranges</span>(histonePeaks))</code></pre></div>
<pre><code>## IRanges object with 6 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1] 141522283 141524219      1937
##   [2]  23560218  23563994      3777
##   [3] 158460103 158464651      4549
##   [4]   1168831   1169616       786
##   [5]  80938636  80944222      5587
##   [6] 159273261 159274505      1245</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># data shifted 5kb to the left</span>
<span class="kw">head</span>(<span class="kw">ranges</span>(shiftedPeaks))</code></pre></div>
<pre><code>## IRanges object with 6 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1] 141517283 141519219      1937
##   [2]  23555218  23558994      3777
##   [3] 158455103 158459651      4549
##   [4]   1163831   1164616       786
##   [5]  80933636  80939222      5587
##   [6] 159268261 159269505      1245</code></pre>
<p>Grab the <code>GC</code> content and plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dnaInShiftedPeaks =<span class="st"> </span><span class="kw">getSeq</span>(BSgenome.Hsapiens.UCSC.hg19,<span class="kw">shift</span>(histonePeaks,<span class="op">-</span><span class="dv">5000</span>))
percentGC1 =<span class="st"> </span><span class="kw">letterFrequency</span>(dnaInShiftedPeaks,<span class="st">&#39;GC&#39;</span>,<span class="dt">as.prob=</span><span class="ot">TRUE</span>)
<span class="kw">plot</span>(<span class="kw">density</span>(percentGC))
<span class="kw">lines</span>(<span class="kw">density</span>(percentGC1),<span class="dt">col=</span><span class="st">&#39;red&#39;</span>)</code></pre></div>
<p><img src="F_epigenomic_ranges_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
</div>
<div id="compare-peak-sets" class="section level3">
<h3><span class="header-section-number">13.7.2</span> Compare peak sets</h3>
<p>As with many datasets in biology, learning about how data relate to each other can be a very powerful technique for hypothesis generation. The goal of the next section is to determine the similarities among a set of 10 histone mark datasets in hopes of generating some hypotheses about the functionality of the histone marks.</p>
<p>Start by loading the datasets that we found above for the K562 cell line.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hpList =<span class="st"> </span><span class="kw">sapply</span>(histoneSets,<span class="cf">function</span>(ahname) ah[[ahname]])
<span class="co"># get the &quot;titles&quot; for the experiment</span>
<span class="co"># And cleanup the messy names</span>
hptitles =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;E123-|</span><span class="ch">\\</span><span class="st">.narrowPeak</span><span class="ch">\\</span><span class="st">.gz&#39;</span>,<span class="st">&#39;&#39;</span>,ah[histoneSets]<span class="op">$</span>title)
<span class="kw">names</span>(hpList) =<span class="st"> </span>hptitles
hptitles</code></pre></div>
<pre><code>##  [1] &quot;H3K4me1&quot;  &quot;H3K4me2&quot;  &quot;H3K4me3&quot;  &quot;H3K9ac&quot;   &quot;H3K9me1&quot;  &quot;H3K9me3&quot;  &quot;H3K27ac&quot;  &quot;H3K27me3&quot;
##  [9] &quot;H3K36me3&quot; &quot;H3K79me2&quot;</code></pre>
<p>How can we compare two sets of regions?</p>
<p>The Jaccard Similarity measure quantifies the proportion of shared peaks (or proportion of bases covered by the shared peaks) to that covered by the <em>union</em> of those peak sets. In mathematical terms, the <em>Jaccard Index</em> measures this similarity.</p>
<p><span class="math display">\[ J(A,B) = \frac{\left | A \bigcap B \right |}{\left | A \bigcup B \right |} \]</span></p>
<p>The <em>Jaccard Distance</em> simply measures how dissimilar two sets of regions are (1-J(A,b)).</p>
<p><span class="math display">\[ d_J(A,B) = 1 - J(A,B) = 1 - \frac{\left | A \bigcap B \right |}{\left | A \bigcup B \right |} \]</span></p>
<p>The <code>regionJaccard()</code> function takes as input two sets of regions and calculates the Jaccard Index using base-level overlaps (as opposed to whole-region overlaps).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&#39; Calculate Jaccard Similarity between two sets of ranges</span>
<span class="co">#&#39;</span>
<span class="co">#&#39; @param a A set of ranges</span>
<span class="co">#&#39; @param b A second set of ranges</span>
<span class="co">#&#39;</span>
<span class="co">#&#39; @details We calculate the Jaccard Distance</span>
<span class="co">#&#39;     as a simple ratio of the number of</span>
<span class="co">#&#39;     shared bases in the regions to</span>
<span class="co">#&#39;     total number of bases covered by</span>
<span class="co">#&#39;     both sets.</span>
<span class="co">#&#39;</span>
<span class="co">#&#39; @return a value between 0 and 1</span>
regionJaccard =<span class="st"> </span><span class="cf">function</span>(a,b) {
    <span class="kw">sum</span>(<span class="kw">width</span>(<span class="kw">intersect</span>(a,b)))<span class="op">/</span><span class="kw">sum</span>(<span class="kw">width</span>(<span class="kw">union</span>(a,b)))
}</code></pre></div>
<p>We can try this on two peakSets:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">regionJaccard</span>(hpList[[<span class="dv">1</span>]],hpList[[<span class="dv">2</span>]])</code></pre></div>
<pre><code>## [1] 0.2811043</code></pre>
<p>How would you interpret the result?</p>
<p>To make addressing the question above a bit more intuitive, we can perform the same calculation, but between all pairs of regions using the <code>multiRegStat()</code> function below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">multiRegStat &lt;-<span class="st"> </span><span class="cf">function</span>(a,b, fun) {
    res &lt;-<span class="st"> </span><span class="kw">matrix</span>(0L, <span class="kw">length</span>(a), <span class="kw">length</span>(b))
    <span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(a))
    {
       <span class="cf">for</span> (j <span class="cf">in</span> <span class="kw">seq_along</span>(b))
       {
          res[i,j] =<span class="st"> </span><span class="kw">fun</span>(a[[i]],b[[j]])
       }
    }
    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.null</span>(<span class="kw">names</span>(a))) <span class="kw">rownames</span>(res) =<span class="st"> </span><span class="kw">names</span>(a)
    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.null</span>(<span class="kw">names</span>(b))) <span class="kw">colnames</span>(res) =<span class="st"> </span><span class="kw">names</span>(b)
    
    res
}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res =<span class="st"> </span><span class="kw">multiRegStat</span>(hpList,hpList,regionJaccard)
res</code></pre></div>
<pre><code>##              H3K4me1     H3K4me2     H3K4me3       H3K9ac      H3K9me1      H3K9me3      H3K27ac
## H3K4me1  1.000000000 0.281104268 0.183791239 0.1927281172 0.0244827526 0.0027769509 0.2452366595
## H3K4me2  0.281104268 1.000000000 0.637133591 0.5686767173 0.0175960031 0.0020254595 0.5097245220
## H3K4me3  0.183791239 0.637133591 1.000000000 0.6514034239 0.0132880861 0.0022354848 0.5049994635
## H3K9ac   0.192728117 0.568676717 0.651403424 1.0000000000 0.0129846178 0.0011958500 0.6090726088
## H3K9me1  0.024482753 0.017596003 0.013288086 0.0129846178 1.0000000000 0.0006955316 0.0167764608
## H3K9me3  0.002776951 0.002025459 0.002235485 0.0011958500 0.0006955316 1.0000000000 0.0039756797
## H3K27ac  0.245236659 0.509724522 0.504999464 0.6090726088 0.0167764608 0.0039756797 1.0000000000
## H3K27me3 0.004769609 0.006489705 0.003637553 0.0007145309 0.0107676075 0.0016215449 0.0008692706
## H3K36me3 0.059625145 0.034442331 0.029895734 0.0308737432 0.0202655273 0.0101396064 0.0417566558
## H3K79me2 0.156555342 0.221234307 0.232402857 0.2325234359 0.0225853902 0.0035804367 0.2030078886
##              H3K27me3     H3K36me3     H3K79me2
## H3K4me1  4.769609e-03 0.0596251450 1.565553e-01
## H3K4me2  6.489705e-03 0.0344423310 2.212343e-01
## H3K4me3  3.637553e-03 0.0298957337 2.324029e-01
## H3K9ac   7.145309e-04 0.0308737432 2.325234e-01
## H3K9me1  1.076761e-02 0.0202655273 2.258539e-02
## H3K9me3  1.621545e-03 0.0101396064 3.580437e-03
## H3K27ac  8.692706e-04 0.0417566558 2.030079e-01
## H3K27me3 1.000000e+00 0.0003409222 8.774185e-05
## H3K36me3 3.409222e-04 1.0000000000 1.170739e-01
## H3K79me2 8.774185e-05 0.1170739250 1.000000e+00</code></pre>
<p>The <code>res</code> matrix now describes the pairwise similarities between the various histone mark peak sets. Making a heatmap and annotating it with knowledge of <a href="https://en.wikipedia.org/wiki/Histone_code">wikipedia, for example</a> is potentially helpful.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ComplexHeatmap)
annot =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">exprStatus=</span><span class="kw">c</span>(<span class="st">&#39;Active&#39;</span>,<span class="st">&#39;Active&#39;</span>,<span class="st">&#39;Active&#39;</span>,<span class="st">&#39;Active&#39;</span>,
                                <span class="st">&#39;Active&#39;</span>,<span class="st">&#39;Repressed&#39;</span>,<span class="st">&#39;Active&#39;</span>,<span class="st">&#39;Repressed&#39;</span>,<span class="ot">NA</span>,<span class="st">&#39;Active&#39;</span>),
                   <span class="dt">locale=</span><span class="kw">c</span>(<span class="ot">NA</span>,<span class="ot">NA</span>,<span class="st">&#39;Promoter&#39;</span>,<span class="st">&#39;Promoter&#39;</span>,<span class="ot">NA</span>,<span class="st">&#39;Body&#39;</span>,<span class="st">&#39;Enhancer&#39;</span>,<span class="st">&#39;Body&#39;</span>,<span class="st">&#39;Body&#39;</span>,<span class="ot">NA</span>),
                   <span class="dt">row.names=</span>hptitles)
<span class="kw">Heatmap</span>(res,<span class="dt">top_annotation=</span><span class="kw">HeatmapAnnotation</span>(
          <span class="dt">df=</span>annot,<span class="dt">col=</span><span class="kw">list</span>(<span class="dt">exprStatus=</span><span class="kw">c</span>(<span class="st">&#39;Active&#39;</span>=<span class="st">&#39;yellow&#39;</span>,<span class="dt">Repressed=</span><span class="st">&#39;blue&#39;</span>))))</code></pre></div>
<p><img src="F_epigenomic_ranges_files/figure-html/unnamed-chunk-34-1.png" width="672" /></p>
<p>What do you think of the plot above?</p>
</div>
<div id="histone-profiles" class="section level3">
<h3><span class="header-section-number">13.7.3</span> Histone profiles</h3>
<p>In the following code block, we:</p>
<ol style="list-style-type: decimal">
<li>Grab the “regions” of “promoters”, defined as a window around the transcript start site.
<ul>
<li>Subsample the promoters randomly to reduce runtime and memory usage.</li>
</ul></li>
<li>Calculate “coverage” of our peaks. For each base in the genome, we will have a single number, mostly zeros, and some ones.</li>
<li>Create slices of our coverage data, called “Views”.</li>
<li>Convert the slices of coverage to rows of a matrix.</li>
<li>Sum the columns of the matrix to produce a summary binding profile.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wndw =<span class="st"> </span><span class="dv">2000</span>
proms =<span class="st"> </span><span class="kw">promoters</span>(txdb, <span class="dt">upstream =</span> wndw, <span class="dt">downstream =</span> wndw,
                  <span class="dt">columns =</span> <span class="kw">c</span>(<span class="st">&#39;tx_name&#39;</span>,<span class="st">&#39;gene_id&#39;</span>))</code></pre></div>
<pre><code>## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 1 out-of-bound range located on sequence GL000199.1. Note that
##   only ranges located on a non-circular sequence whose length is not NA can be considered
##   out-of-bound (use seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges. See
##   ?`trim,GenomicRanges-method` for more information.

## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 1 out-of-bound range located on sequence GL000199.1. Note that
##   only ranges located on a non-circular sequence whose length is not NA can be considered
##   out-of-bound (use seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges. See
##   ?`trim,GenomicRanges-method` for more information.

## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 1 out-of-bound range located on sequence GL000199.1. Note that
##   only ranges located on a non-circular sequence whose length is not NA can be considered
##   out-of-bound (use seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges. See
##   ?`trim,GenomicRanges-method` for more information.</code></pre>
<pre><code>## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 2 out-of-bound ranges located on sequences GL000199.1 and
##   GL000240.1. Note that only ranges located on a non-circular sequence whose length is not
##   NA can be considered out-of-bound (use seqlengths() and isCircular() to get the lengths
##   and circularity flags of the underlying sequences). You can use trim() to trim these
##   ranges. See ?`trim,GenomicRanges-method` for more information.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">proms =<span class="st"> </span>proms[<span class="kw">sample</span>(<span class="kw">seq_along</span>(proms),<span class="dv">20000</span>,<span class="dt">replace=</span><span class="ot">FALSE</span>)]
proms =<span class="st"> </span><span class="kw">keepStandardChromosomes</span>(proms,<span class="dt">pruning.mode=</span><span class="st">&#39;tidy&#39;</span>)
covg =<span class="st"> </span><span class="kw">coverage</span>(hpList[[<span class="dv">1</span>]])
<span class="kw">seqlevelsStyle</span>(proms) =<span class="st"> </span><span class="kw">seqlevelsStyle</span>(covg)
covg =<span class="st"> </span>covg[<span class="kw">seqlevels</span>(proms)]
vi =<span class="st"> </span><span class="kw">Views</span>(covg,<span class="kw">as</span>(proms[<span class="kw">strand</span>(proms)<span class="op">==</span><span class="st">&#39;+&#39;</span>,],<span class="st">&#39;RangesList&#39;</span>))
m =<span class="st"> </span><span class="kw">as.matrix</span>(vi)
<span class="kw">plot</span>(<span class="kw">colSums</span>(m,<span class="dt">na.rm=</span><span class="ot">TRUE</span>),<span class="dt">type=</span><span class="st">&#39;l&#39;</span>)</code></pre></div>
<p><img src="F_epigenomic_ranges_files/figure-html/unnamed-chunk-35-1.png" width="672" /></p>
<p>We can wrap all this up as a function to reuse as we see fit.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">histoneProfile =<span class="st"> </span><span class="cf">function</span>(promoters,peaks,<span class="dt">n_promoters=</span><span class="dv">10000</span>) {
    proms =<span class="st"> </span>promoters
    proms =<span class="st"> </span>proms[<span class="kw">sample</span>(<span class="kw">seq_along</span>(proms),<span class="kw">min</span>(<span class="kw">length</span>(proms),n_promoters),<span class="dt">replace=</span><span class="ot">FALSE</span>)]
    proms =<span class="st"> </span><span class="kw">keepStandardChromosomes</span>(proms,<span class="dt">pruning.mode=</span><span class="st">&#39;tidy&#39;</span>)
    covg =<span class="st"> </span><span class="kw">coverage</span>(peaks)
    <span class="kw">seqlevelsStyle</span>(proms) =<span class="st"> </span><span class="kw">seqlevelsStyle</span>(covg)
    covg =<span class="st"> </span>covg[<span class="kw">seqlevels</span>(proms)]
    vi =<span class="st"> </span><span class="kw">Views</span>(covg,<span class="kw">as</span>(proms[<span class="kw">strand</span>(proms)<span class="op">==</span><span class="st">&#39;+&#39;</span>,],<span class="st">&#39;RangesList&#39;</span>))
    <span class="kw">return</span>(<span class="kw">colSums</span>(<span class="kw">as.matrix</span>(vi),<span class="dt">na.rm =</span> <span class="ot">TRUE</span>))
}</code></pre></div>
<p>Now, we are going to load gene expression for K562 so that we can look at how the binding profiles vary with expression. The gene expression from ENCODE data is available from many sources. In this case, we are downloading directly from Ensembl in Europe.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(readr)
k562expr =<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&#39;https://www.encodeproject.org/files/ENCFF812ADD/@@download/ENCFF812ADD.tsv&#39;</span>)</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   gene_id = col_character(),
##   `transcript_id(s)` = col_character(),
##   length = col_double(),
##   effective_length = col_double(),
##   expected_count = col_double(),
##   TPM = col_double(),
##   FPKM = col_double(),
##   posterior_mean_count = col_double(),
##   posterior_standard_deviation_of_count = col_double(),
##   pme_TPM = col_double(),
##   pme_FPKM = col_double(),
##   TPM_ci_lower_bound = col_double(),
##   TPM_ci_upper_bound = col_double(),
##   FPKM_ci_lower_bound = col_double(),
##   FPKM_ci_upper_bound = col_double()
## )</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(k562expr)</code></pre></div>
<pre><code>## # A tibble: 6 x 15
##   gene_id `transcript_id(s)` length effective_length expected_count   TPM  FPKM
##     &lt;chr&gt;              &lt;chr&gt;  &lt;dbl&gt;            &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1   10904              10904     93               60           0.00  0.00  0.00
## 2   12954              12954     94               61           0.00  0.00  0.00
## 3   12956              12956     72               39           0.00  0.00  0.00
## 4   12958              12958     82               49          22.22 18.58 17.10
## 5   12960              12960     73               40           0.00  0.00  0.00
## 6   12962              12962     72               39           1.00  1.05  0.96
## # ... with 8 more variables: posterior_mean_count &lt;dbl&gt;,
## #   posterior_standard_deviation_of_count &lt;dbl&gt;, pme_TPM &lt;dbl&gt;, pme_FPKM &lt;dbl&gt;,
## #   TPM_ci_lower_bound &lt;dbl&gt;, TPM_ci_upper_bound &lt;dbl&gt;, FPKM_ci_lower_bound &lt;dbl&gt;,
## #   FPKM_ci_upper_bound &lt;dbl&gt;</code></pre>
<p>These data have several columns of values, including <code>tpm</code> or transcripts-per-million, which we be using as our measure of expression. Go ahead and take a look at the distribution and other characteristics of the <code>tpm</code> column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fix gene names</span>
k562expr<span class="op">$</span>gene_id=<span class="kw">sub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">..*&#39;</span>,<span class="st">&#39;&#39;</span>,k562expr<span class="op">$</span>gene_id)
k562expr =<span class="st"> </span><span class="kw">subset</span>(k562expr,<span class="kw">grepl</span>(<span class="st">&#39;ENSG&#39;</span>,gene_id))</code></pre></div>
<p>Below, I am simply attaching the expression values to their associated promoters based on gene id matching.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">proms<span class="op">$</span>TPM =<span class="st"> </span>k562expr<span class="op">$</span>TPM[<span class="kw">match</span>(<span class="kw">unlist</span>(proms<span class="op">$</span>gene_id),k562expr<span class="op">$</span>gene_id)]
proms<span class="op">$</span>logTPM =
<span class="st">    </span><span class="kw">log10</span>(k562expr<span class="op">$</span>TPM[<span class="kw">match</span>(<span class="kw">unlist</span>(proms<span class="op">$</span>gene_id),k562expr<span class="op">$</span>gene_id)])
<span class="kw">head</span>(proms)</code></pre></div>
<pre><code>## GRanges object with 6 ranges and 4 metadata columns:
##       seqnames                 ranges strand |         tx_name         gene_id       TPM     logTPM
##          &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;CharacterList&gt; &lt;numeric&gt;  &lt;numeric&gt;
##   [1]    chr16 [ 11648301,  11652300]      - | ENST00000575426 ENSG00000189067      4.78  0.6794279
##   [2]    chr19 [  3918505,   3922504]      - | ENST00000584410 ENSG00000266627      0.00       -Inf
##   [3]     chrX [145894245, 145898244]      - | ENST00000458472 ENSG00000224440      0.00       -Inf
##   [4]    chr15 [ 79269468,  79273467]      - | ENST00000559926 ENSG00000058335      0.03 -1.5228787
##   [5]     chr4 [ 52774919,  52778918]      + | ENST00000510518 ENSG00000109184     15.79  1.1983821
##   [6]     chr1 [213173577, 213177576]      - | ENST00000473303 ENSG00000174606     12.34  1.0913152
##   -------
##   seqinfo: 25 sequences (1 circular) from GRCh37 genome</code></pre>
<p>Now, subset promoters into those with low expression and those with high.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">promslow =<span class="st"> </span>proms[<span class="kw">which</span>(proms<span class="op">$</span>logTPM<span class="op">&lt;</span><span class="kw">median</span>(proms<span class="op">$</span>logTPM,<span class="dt">na.rm=</span><span class="ot">TRUE</span>))]
promshigh =<span class="st"> </span>proms[<span class="kw">which</span>(proms<span class="op">$</span>logTPM<span class="op">&gt;=</span><span class="kw">median</span>(proms<span class="op">$</span>logTPM,<span class="dt">na.rm=</span><span class="ot">TRUE</span>))]</code></pre></div>
<p>And finally, produce a plot of two histone marks with low and high expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat =<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">histoneProfile</span>(promslow,hpList[[<span class="dv">1</span>]]),
            <span class="kw">histoneProfile</span>(promshigh,hpList[[<span class="dv">1</span>]]),
            <span class="kw">histoneProfile</span>(promslow,hpList[[<span class="dv">10</span>]]),
            <span class="kw">histoneProfile</span>(promshigh,hpList[[<span class="dv">10</span>]]))
<span class="kw">matplot</span>(mat,<span class="dt">type=</span><span class="st">&#39;l&#39;</span>,<span class="dt">main=</span><span class="kw">paste</span>(hptitles[<span class="dv">1</span>],hptitles[[<span class="dv">10</span>]], <span class="dt">sep=</span><span class="st">&quot; &amp; &quot;</span>))
<span class="kw">legend</span>(<span class="dv">0</span>,<span class="dv">2000</span>,<span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&#39;H3K4me1_low&#39;</span>,<span class="st">&#39;H3K4me1_high&#39;</span>,<span class="st">&#39;H3K79me2_low&#39;</span>,<span class="st">&#39;H3K79me2_high&#39;</span>),
       <span class="dt">col=</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>,<span class="dt">lty=</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)</code></pre></div>
<p><img src="F_epigenomic_ranges_files/figure-html/unnamed-chunk-41-1.png" width="672" /></p>
</div>
</div>
<div id="sessioninfo" class="section level2">
<h2><span class="header-section-number">13.8</span> <code>sessionInfo()</code></h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R Under development (unstable) (2016-10-26 r71594)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Sierra 10.12.4
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 GenomicAlignments_1.11.11              
##  [3] Rsamtools_1.27.13                       SummarizedExperiment_1.5.9             
##  [5] DelayedArray_0.1.11                     matrixStats_0.52.2                     
##  [7] GenomicFeatures_1.27.9                  AnnotationDbi_1.37.4                   
##  [9] Biobase_2.35.1                          readr_1.1.1                            
## [11] BSgenome.Hsapiens.UCSC.hg19_1.4.0       BSgenome_1.43.7                        
## [13] Biostrings_2.43.8                       XVector_0.15.2                         
## [15] rtracklayer_1.35.6                      GenomicRanges_1.27.23                  
## [17] GenomeInfoDb_1.11.11                    IRanges_2.9.19                         
## [19] S4Vectors_0.13.17                       shiny_1.0.1                            
## [21] AnnotationHub_2.7.14                    BiocGenerics_0.21.3                    
## [23] plotly_4.7.0                            ggplot2_2.2.1.9000                     
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.2.1                    tidyr_0.6.1                   jsonlite_1.4                 
##  [4] viridisLite_0.2.0             assertthat_0.2.0              interactiveDisplayBase_1.13.0
##  [7] GenomeInfoDbData_0.99.0       yaml_2.1.14                   RSQLite_1.1-2                
## [10] backports_1.0.5               lattice_0.20-35               glue_1.1.1                   
## [13] digest_0.6.12                 colorspace_1.3-2              htmltools_0.3.5              
## [16] httpuv_1.3.3                  Matrix_1.2-8                  plyr_1.8.4                   
## [19] XML_3.98-1.6                  pkgconfig_2.0.1               biomaRt_2.32.1               
## [22] bookdown_0.4.1                zlibbioc_1.21.0               purrr_0.2.2                  
## [25] xtable_1.8-2                  scales_0.4.1                  BiocParallel_1.9.6           
## [28] tibble_1.3.3                  lazyeval_0.2.0                magrittr_1.5                 
## [31] mime_0.5                      memoise_1.1.0                 evaluate_0.10                
## [34] BiocInstaller_1.26.0          tools_3.4.0                   data.table_1.10.4            
## [37] hms_0.3                       stringr_1.2.0                 munsell_0.4.3                
## [40] bindrcpp_0.2                  compiler_3.4.0                rlang_0.1.1                  
## [43] grid_3.4.0                    RCurl_1.95-4.8                htmlwidgets_0.8              
## [46] crosstalk_1.0.0               bitops_1.0-6                  labeling_0.3                 
## [49] rmarkdown_1.4                 codetools_0.2-15              gtable_0.2.0                 
## [52] DBI_0.6-1                     curl_2.5                      R6_2.2.0                     
## [55] knitr_1.16                    dplyr_0.7.1                   bindr_0.1                    
## [58] rprojroot_1.2                 stringi_1.1.5                 Rcpp_0.12.11</code></pre>

<p>lsls— author: “Sean Davis” date: “6/29/2017” output: html_document: default pdf_document: default —</p>
</div>
</div>



    <script>
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
        ga('create', 'UA-93043521-1', 'auto');
        ga('send', 'pageview');

         
        var links = document.querySelectorAll('a');
        Array.prototype.map.call(links, function(item) {
            if (item.host != document.location.host) {
                item.addEventListener('click', function() {
                    var action = item.getAttribute('data-action') || 'follow';
                    ga('send', 'event', 'outbound', action, item.href);
                });
            }
        });
    </script>
            </section>

          </div>
        </div>
      </div>
<a href="graphics-and-visualization.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="appendix-data-sets.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="book_assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="book_assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="book_assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/seandavi/ITR/edit/master/F_epigenomic_ranges.Rmd",
"text": "Edit"
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
